From a4e73da63c431db90357aed23602ca5f66bdd89b Mon Sep 17 00:00:00 2001
From: Sven van Haastregt <sven.vanhaastregt@arm.com>
Date: Fri, 20 May 2022 17:34:02 +0100
Subject: [PATCH 044/172] [NFC] Move pass declarations into separate header
 files

Prepare for migrating SPIRVWriter to the new PassManager by making
sure the new non-legacy pass declarations are visible outside of their
.cpp file.

(cherry picked from commit 337b171131b13487e280752c92508ded4291aff6)
---
 lib/SPIRV/OCLToSPIRV.cpp                      | 333 +++---------------
 lib/SPIRV/OCLToSPIRV.h                        | 296 ++++++++++++++++
 lib/SPIRV/OCLTypeToSPIRV.cpp                  |   2 +
 lib/SPIRV/PreprocessMetadata.cpp              |  39 +-
 lib/SPIRV/PreprocessMetadata.h                |  90 +++++
 .../SPIRVLowerBitCastToNonStandardType.cpp    | 221 ++++++------
 .../SPIRVLowerBitCastToNonStandardType.h      |  79 +++++
 lib/SPIRV/SPIRVLowerBool.cpp                  | 176 +++++----
 lib/SPIRV/SPIRVLowerBool.h                    |  87 +++++
 lib/SPIRV/SPIRVLowerMemmove.cpp               | 182 +++++-----
 lib/SPIRV/SPIRVLowerMemmove.h                 |  81 +++++
 lib/SPIRV/SPIRVLowerOCLBlocks.cpp             |  65 ++--
 lib/SPIRV/SPIRVLowerOCLBlocks.h               |  74 ++++
 lib/SPIRV/SPIRVLowerSaddWithOverflow.cpp      | 164 ++++-----
 lib/SPIRV/SPIRVLowerSaddWithOverflow.h        |  81 +++++
 lib/SPIRV/SPIRVRegularizeLLVM.cpp             | 100 +-----
 lib/SPIRV/SPIRVRegularizeLLVM.h               | 143 ++++++++
 17 files changed, 1356 insertions(+), 857 deletions(-)
 create mode 100644 lib/SPIRV/OCLToSPIRV.h
 create mode 100644 lib/SPIRV/PreprocessMetadata.h
 create mode 100644 lib/SPIRV/SPIRVLowerBitCastToNonStandardType.h
 create mode 100644 lib/SPIRV/SPIRVLowerBool.h
 create mode 100644 lib/SPIRV/SPIRVLowerMemmove.h
 create mode 100644 lib/SPIRV/SPIRVLowerOCLBlocks.h
 create mode 100644 lib/SPIRV/SPIRVLowerSaddWithOverflow.h
 create mode 100644 lib/SPIRV/SPIRVRegularizeLLVM.h

diff --git a/lib/SPIRV/OCLToSPIRV.cpp b/lib/SPIRV/OCLToSPIRV.cpp
index 6ebc4bfd..9a64fdc0 100644
--- a/lib/SPIRV/OCLToSPIRV.cpp
+++ b/lib/SPIRV/OCLToSPIRV.cpp
@@ -38,19 +38,16 @@
 //===----------------------------------------------------------------------===//
 #define DEBUG_TYPE "ocl-to-spv"
 
+#include "OCLToSPIRV.h"
 #include "OCLTypeToSPIRV.h"
-#include "OCLUtil.h"
 #include "SPIRVInternal.h"
 #include "libSPIRV/SPIRVDebug.h"
 
 #include "llvm/ADT/StringSwitch.h"
 #include "llvm/Analysis/ValueTracking.h"
 #include "llvm/IR/IRBuilder.h"
-#include "llvm/IR/InstVisitor.h"
 #include "llvm/IR/Instruction.h"
 #include "llvm/IR/Instructions.h"
-#include "llvm/IR/PassManager.h"
-#include "llvm/Pass.h"
 #include "llvm/Support/Debug.h"
 
 #include <algorithm>
@@ -70,290 +67,64 @@ static size_t getOCLCpp11AtomicMaxNumOps(StringRef Name) {
       .Default(0);
 }
 
-class OCLToSPIRVBase : public InstVisitor<OCLToSPIRVBase> {
-public:
-  OCLToSPIRVBase() : M(nullptr), Ctx(nullptr), CLVer(0) {}
-  virtual ~OCLToSPIRVBase() {}
-  bool runOCLToSPIRV(Module &M);
-
-  virtual void visitCallInst(CallInst &CI);
-
-  /// Transform barrier/work_group_barrier/sub_group_barrier
-  ///     to __spirv_ControlBarrier.
-  /// barrier(flag) =>
-  ///   __spirv_ControlBarrier(workgroup, workgroup, map(flag))
-  /// work_group_barrier(scope, flag) =>
-  ///   __spirv_ControlBarrier(workgroup, map(scope), map(flag))
-  /// sub_group_barrier(scope, flag) =>
-  ///   __spirv_ControlBarrier(subgroup, map(scope), map(flag))
-  void visitCallBarrier(CallInst *CI);
-
-  /// Erase useless convert functions.
-  /// \return true if the call instruction is erased.
-  bool eraseUselessConvert(CallInst *Call, StringRef MangledName,
-                           StringRef DeMangledName);
-
-  /// Transform convert_ to
-  ///   __spirv_{CastOpName}_R{TargeTyName}{_sat}{_rt[p|n|z|e]}
-  void visitCallConvert(CallInst *CI, StringRef MangledName,
-                        StringRef DemangledName);
-
-  /// Transform async_work_group{_strided}_copy.
-  /// async_work_group_copy(dst, src, n, event)
-  ///   => async_work_group_strided_copy(dst, src, n, 1, event)
-  /// async_work_group_strided_copy(dst, src, n, stride, event)
-  ///   => __spirv_AsyncGroupCopy(ScopeWorkGroup, dst, src, n, stride, event)
-  void visitCallAsyncWorkGroupCopy(CallInst *CI, StringRef DemangledName);
-
-  /// Transform OCL builtin function to SPIR-V builtin function.
-  void transBuiltin(CallInst *CI, OCLBuiltinTransInfo &Info);
-
-  /// Transform atomic_work_item_fence/mem_fence to __spirv_MemoryBarrier.
-  /// func(flag, order, scope) =>
-  ///   __spirv_MemoryBarrier(map(scope), map(flag)|map(order))
-  void transMemoryBarrier(CallInst *CI, AtomicWorkItemFenceLiterals);
-
-  /// Transform all to __spirv_Op(All|Any).  Note that the types mismatch so
-  // some extra code is emitted to convert between the two.
-  void visitCallAllAny(spv::Op OC, CallInst *CI);
-
-  /// Transform atomic_* to __spirv_Atomic*.
-  /// atomic_x(ptr_arg, args, order, scope) =>
-  ///   __spirv_AtomicY(ptr_arg, map(order), map(scope), args)
-  void transAtomicBuiltin(CallInst *CI, OCLBuiltinTransInfo &Info);
-
-  /// Transform atomic_work_item_fence to __spirv_MemoryBarrier.
-  /// atomic_work_item_fence(flag, order, scope) =>
-  ///   __spirv_MemoryBarrier(map(scope), map(flag)|map(order))
-  void visitCallAtomicWorkItemFence(CallInst *CI);
-
-  /// Transform atomic_compare_exchange call.
-  /// In atomic_compare_exchange, the expected value parameter is a pointer.
-  /// However in SPIR-V it is a value. The transformation adds a load
-  /// instruction, result of which is passed to atomic_compare_exchange as
-  /// argument.
-  /// The transformation adds a store instruction after the call, to update the
-  /// value in expected with the value pointed to by object. Though, it is not
-  /// necessary in case they are equal, this approach makes result code simpler.
-  /// Also ICmp instruction is added, because the call must return result of
-  /// comparison.
-  /// \returns the call instruction of atomic_compare_exchange_strong.
-  CallInst *visitCallAtomicCmpXchg(CallInst *CI);
-
-  /// Transform atomic_init.
-  /// atomic_init(p, x) => store p, x
-  void visitCallAtomicInit(CallInst *CI);
-
-  /// Transform legacy OCL 1.x atomic builtins to SPIR-V builtins for extensions
-  ///   cl_khr_int64_base_atomics
-  ///   cl_khr_int64_extended_atomics
-  /// Do nothing if the called function is not a legacy atomic builtin.
-  void visitCallAtomicLegacy(CallInst *CI, StringRef MangledName,
-                             StringRef DemangledName);
-
-  /// Transform OCL 2.0 C++11 atomic builtins to SPIR-V builtins.
-  /// Do nothing if the called function is not a C++11 atomic builtin.
-  void visitCallAtomicCpp11(CallInst *CI, StringRef MangledName,
-                            StringRef DemangledName);
-
-  /// Transform OCL builtin function to SPIR-V builtin function.
-  /// Assuming there is a simple name mapping without argument changes.
-  /// Should be called at last.
-  void visitCallBuiltinSimple(CallInst *CI, StringRef MangledName,
-                              StringRef DemangledName);
-
-  /// Transform get_image_{width|height|depth|dim}.
-  /// get_image_xxx(...) =>
-  ///   dimension = __spirv_ImageQuerySizeLod_R{ReturnType}(...);
-  ///   return dimension.{x|y|z};
-  void visitCallGetImageSize(CallInst *CI, StringRef DemangledName);
-
-  /// Transform {work|sub}_group_x =>
-  ///   __spirv_{OpName}
-  ///
-  /// Special handling of work_group_broadcast.
-  ///   work_group_broadcast(a, x, y, z)
-  ///     =>
-  ///   __spirv_GroupBroadcast(a, vec3(x, y, z))
-
-  void visitCallGroupBuiltin(CallInst *CI, StringRef DemangledName);
-
-  /// Transform mem_fence to __spirv_MemoryBarrier.
-  /// mem_fence(flag) => __spirv_MemoryBarrier(Workgroup, map(flag))
-  void visitCallMemFence(CallInst *CI, StringRef DemangledName);
-
-  void visitCallNDRange(CallInst *CI, StringRef DemangledName);
-
-  /// Transform read_image with sampler arguments.
-  /// read_image(image, sampler, ...) =>
-  ///   sampled_image = __spirv_SampledImage(image, sampler);
-  ///   return __spirv_ImageSampleExplicitLod_R{ReturnType}(sampled_image, ...);
-  void visitCallReadImageWithSampler(CallInst *CI, StringRef MangledName);
-
-  /// Transform read_image with msaa image arguments.
-  /// Sample argument must be acoded as Image Operand.
-  void visitCallReadImageMSAA(CallInst *CI, StringRef MangledName);
-
-  /// Transform {read|write}_image without sampler arguments.
-  void visitCallReadWriteImage(CallInst *CI, StringRef DemangledName);
-
-  /// Transform to_{global|local|private}.
-  ///
-  /// T* a = ...;
-  /// addr T* b = to_addr(a);
-  ///   =>
-  /// i8* x = cast<i8*>(a);
-  /// addr i8* y = __spirv_GenericCastToPtr_ToAddr(x);
-  /// addr T* b = cast<addr T*>(y);
-  void visitCallToAddr(CallInst *CI, StringRef DemangledName);
-
-  /// Transform return type of relatinal built-in functions like isnan, isfinite
-  /// to boolean values.
-  void visitCallRelational(CallInst *CI, StringRef DemangledName);
-
-  /// Transform vector load/store functions to SPIR-V extended builtin
-  ///   functions
-  /// {vload|vstore{a}}{_half}{n}{_rte|_rtz|_rtp|_rtn} =>
-  ///   __spirv_ocl_{ExtendedInstructionOpCodeName}__R{ReturnType}
-  void visitCallVecLoadStore(CallInst *CI, StringRef MangledName,
-                             StringRef DemangledName);
-
-  /// Transforms get_mem_fence built-in to SPIR-V function and aligns result
-  /// values with SPIR 1.2. get_mem_fence(ptr) => __spirv_GenericPtrMemSemantics
-  /// GenericPtrMemSemantics valid values are 0x100, 0x200 and 0x300, where is
-  /// SPIR 1.2 defines them as 0x1, 0x2 and 0x3, so this function adjusts
-  /// GenericPtrMemSemantics results to SPIR 1.2 values.
-  void visitCallGetFence(CallInst *CI, StringRef DemangledName);
-
-  /// Transforms OpDot instructions with a scalar type to a fmul instruction
-  void visitCallDot(CallInst *CI);
-
-  /// Fixes for built-in functions with vector+scalar arguments that are
-  /// translated to the SPIR-V instructions where all arguments must have the
-  /// same type.
-  void visitCallScalToVec(CallInst *CI, StringRef MangledName,
-                          StringRef DemangledName);
-
-  /// Transform get_image_channel_{order|data_type} built-in functions to
-  ///   __spirv_ocl_{ImageQueryOrder|ImageQueryFormat}
-  void visitCallGetImageChannel(CallInst *CI, StringRef DemangledName,
-                                unsigned int Offset);
-
-  /// Transform enqueue_kernel and kernel query built-in functions to
-  /// spirv-friendly format filling arguments, required for device-side enqueue
-  /// instructions, but missed in the original call
-  void visitCallEnqueueKernel(CallInst *CI, StringRef DemangledName);
-  void visitCallKernelQuery(CallInst *CI, StringRef DemangledName);
-
-  /// For cl_intel_subgroups block read built-ins:
-  void visitSubgroupBlockReadINTEL(CallInst *CI);
-
-  /// For cl_intel_subgroups block write built-ins:
-  void visitSubgroupBlockWriteINTEL(CallInst *CI);
-
-  /// For cl_intel_media_block_io built-ins:
-  void visitSubgroupImageMediaBlockINTEL(CallInst *CI, StringRef DemangledName);
-  // For cl_intel_device_side_avc_motion_estimation built-ins
-  void visitSubgroupAVCBuiltinCall(CallInst *CI, StringRef DemangledName);
-  void visitSubgroupAVCWrapperBuiltinCall(CallInst *CI, Op WrappedOC,
-                                          StringRef DemangledName);
-  void visitSubgroupAVCBuiltinCallWithSampler(CallInst *CI,
-                                              StringRef DemangledName);
-
-  void visitCallLdexp(CallInst *CI, StringRef MangledName,
-                      StringRef DemangledName);
-
-  /// For cl_intel_convert_bfloat16_as_ushort
-  void visitCallConvertBFloat16AsUshort(CallInst *CI, StringRef DemangledName);
-  /// For cl_intel_convert_as_bfloat16_float
-  void visitCallConvertAsBFloat16Float(CallInst *CI, StringRef DemangledName);
-
-  void setOCLTypeToSPIRV(OCLTypeToSPIRVBase *OCLTypeToSPIRV) {
-    OCLTypeToSPIRVPtr = OCLTypeToSPIRV;
-  }
-  OCLTypeToSPIRVBase *getOCLTypeToSPIRV() { return OCLTypeToSPIRVPtr; }
-
-private:
-  Module *M;
-  LLVMContext *Ctx;
-  unsigned CLVer; /// OpenCL version as major*10+minor
-  std::set<Value *> ValuesToDelete;
-  OCLTypeToSPIRVBase *OCLTypeToSPIRVPtr;
-
-  ConstantInt *addInt32(int I) { return getInt32(M, I); }
-  ConstantInt *addSizet(uint64_t I) { return getSizet(M, I); }
-
-  /// Get vector width from OpenCL vload* function name.
-  SPIRVWord getVecLoadWidth(const std::string &DemangledName) {
-    SPIRVWord Width = 0;
-    if (DemangledName == "vloada_half")
-      Width = 1;
-    else {
-      unsigned Loc = 5;
-      if (DemangledName.find("vload_half") == 0)
-        Loc = 10;
-      else if (DemangledName.find("vloada_half") == 0)
-        Loc = 11;
-
-      std::stringstream SS(DemangledName.substr(Loc));
-      SS >> Width;
-    }
-    return Width;
-  }
+bool OCLToSPIRVLegacy::runOnModule(Module &M) {
+  setOCLTypeToSPIRV(&getAnalysis<OCLTypeToSPIRVLegacy>());
+  return runOCLToSPIRV(M);
+}
 
-  /// Transform OpenCL vload/vstore function name.
-  void transVecLoadStoreName(std::string &DemangledName,
-                             const std::string &Stem, bool AlwaysN) {
-    auto HalfStem = Stem + "_half";
-    auto HalfStemR = HalfStem + "_r";
-    if (!AlwaysN && DemangledName == HalfStem)
-      return;
-    if (!AlwaysN && DemangledName.find(HalfStemR) == 0) {
-      DemangledName = HalfStemR;
-      return;
-    }
-    if (DemangledName.find(HalfStem) == 0) {
-      auto OldName = DemangledName;
-      DemangledName = HalfStem + "n";
-      if (OldName.find("_r") != std::string::npos)
-        DemangledName += "_r";
-      return;
-    }
-    if (DemangledName.find(Stem) == 0) {
-      DemangledName = Stem + "n";
-      return;
-    }
-  }
-};
+void OCLToSPIRVLegacy::getAnalysisUsage(AnalysisUsage &AU) const {
+  AU.addRequired<OCLTypeToSPIRVLegacy>();
+}
 
-class OCLToSPIRVLegacy : public OCLToSPIRVBase, public llvm::ModulePass {
-public:
-  OCLToSPIRVLegacy() : ModulePass(ID) {
-    initializeOCLToSPIRVLegacyPass(*PassRegistry::getPassRegistry());
-  }
+llvm::PreservedAnalyses OCLToSPIRVPass::run(llvm::Module &M,
+                                            llvm::ModuleAnalysisManager &MAM) {
+  setOCLTypeToSPIRV(&MAM.getResult<OCLTypeToSPIRVPass>(M));
+  return runOCLToSPIRV(M) ? llvm::PreservedAnalyses::none()
+                          : llvm::PreservedAnalyses::all();
+}
 
-  bool runOnModule(Module &M) override {
-    setOCLTypeToSPIRV(&getAnalysis<OCLTypeToSPIRVLegacy>());
-    return runOCLToSPIRV(M);
+/// Get vector width from OpenCL vload* function name.
+SPIRVWord OCLToSPIRVBase::getVecLoadWidth(const std::string &DemangledName) {
+  SPIRVWord Width = 0;
+  if (DemangledName == "vloada_half")
+    Width = 1;
+  else {
+    unsigned Loc = 5;
+    if (DemangledName.find("vload_half") == 0)
+      Loc = 10;
+    else if (DemangledName.find("vloada_half") == 0)
+      Loc = 11;
+
+    std::stringstream SS(DemangledName.substr(Loc));
+    SS >> Width;
   }
+  return Width;
+}
 
-  void getAnalysisUsage(AnalysisUsage &AU) const override {
-    AU.addRequired<OCLTypeToSPIRVLegacy>();
+/// Transform OpenCL vload/vstore function name.
+void OCLToSPIRVBase::transVecLoadStoreName(std::string &DemangledName,
+                                           const std::string &Stem,
+                                           bool AlwaysN) {
+  auto HalfStem = Stem + "_half";
+  auto HalfStemR = HalfStem + "_r";
+  if (!AlwaysN && DemangledName == HalfStem)
+    return;
+  if (!AlwaysN && DemangledName.find(HalfStemR) == 0) {
+    DemangledName = HalfStemR;
+    return;
   }
-
-  static char ID;
-};
-
-class OCLToSPIRVPass : public OCLToSPIRVBase,
-                       public llvm::PassInfoMixin<OCLToSPIRVBase> {
-public:
-  llvm::PreservedAnalyses run(llvm::Module &M,
-                              llvm::ModuleAnalysisManager &MAM) {
-    setOCLTypeToSPIRV(&MAM.getResult<OCLTypeToSPIRVPass>(M));
-    return runOCLToSPIRV(M) ? llvm::PreservedAnalyses::none()
-                            : llvm::PreservedAnalyses::all();
+  if (DemangledName.find(HalfStem) == 0) {
+    auto OldName = DemangledName;
+    DemangledName = HalfStem + "n";
+    if (OldName.find("_r") != std::string::npos)
+      DemangledName += "_r";
+    return;
   }
-};
+  if (DemangledName.find(Stem) == 0) {
+    DemangledName = Stem + "n";
+    return;
+  }
+}
 
 char OCLToSPIRVLegacy::ID = 0;
 
diff --git a/lib/SPIRV/OCLToSPIRV.h b/lib/SPIRV/OCLToSPIRV.h
new file mode 100644
index 00000000..7f80678a
--- /dev/null
+++ b/lib/SPIRV/OCLToSPIRV.h
@@ -0,0 +1,296 @@
+//=- OCLToSPIRV.h - OpenCL to SPIR-V builtin preprocessing pass -*- C++ -*-=//
+//
+//                     The LLVM/SPIR-V Translator
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright (c) 2022 The Khronos Group Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of The Khronos Group, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements preprocessing of OpenCL C built-in functions into SPIR-V
+// friendly IR form for further translation into SPIR-V
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SPIRV_OCLTOSPIRV_H
+#define SPIRV_OCLTOSPIRV_H
+
+#include "OCLUtil.h"
+
+#include "llvm/IR/InstVisitor.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+namespace SPIRV {
+
+class OCLTypeToSPIRVBase;
+
+class OCLToSPIRVBase : public InstVisitor<OCLToSPIRVBase> {
+public:
+  OCLToSPIRVBase() : M(nullptr), Ctx(nullptr), CLVer(0) {}
+  virtual ~OCLToSPIRVBase() {}
+  bool runOCLToSPIRV(Module &M);
+
+  virtual void visitCallInst(CallInst &CI);
+
+  /// Transform barrier/work_group_barrier/sub_group_barrier
+  ///     to __spirv_ControlBarrier.
+  /// barrier(flag) =>
+  ///   __spirv_ControlBarrier(workgroup, workgroup, map(flag))
+  /// work_group_barrier(scope, flag) =>
+  ///   __spirv_ControlBarrier(workgroup, map(scope), map(flag))
+  /// sub_group_barrier(scope, flag) =>
+  ///   __spirv_ControlBarrier(subgroup, map(scope), map(flag))
+  void visitCallBarrier(CallInst *CI);
+
+  /// Erase useless convert functions.
+  /// \return true if the call instruction is erased.
+  bool eraseUselessConvert(CallInst *Call, StringRef MangledName,
+                           StringRef DeMangledName);
+
+  /// Transform convert_ to
+  ///   __spirv_{CastOpName}_R{TargeTyName}{_sat}{_rt[p|n|z|e]}
+  void visitCallConvert(CallInst *CI, StringRef MangledName,
+                        StringRef DemangledName);
+
+  /// Transform async_work_group{_strided}_copy.
+  /// async_work_group_copy(dst, src, n, event)
+  ///   => async_work_group_strided_copy(dst, src, n, 1, event)
+  /// async_work_group_strided_copy(dst, src, n, stride, event)
+  ///   => __spirv_AsyncGroupCopy(ScopeWorkGroup, dst, src, n, stride, event)
+  void visitCallAsyncWorkGroupCopy(CallInst *CI, StringRef DemangledName);
+
+  /// Transform OCL builtin function to SPIR-V builtin function.
+  void transBuiltin(CallInst *CI, OCLBuiltinTransInfo &Info);
+
+  /// Transform atomic_work_item_fence/mem_fence to __spirv_MemoryBarrier.
+  /// func(flag, order, scope) =>
+  ///   __spirv_MemoryBarrier(map(scope), map(flag)|map(order))
+  void transMemoryBarrier(CallInst *CI, AtomicWorkItemFenceLiterals);
+
+  /// Transform all to __spirv_Op(All|Any).  Note that the types mismatch so
+  // some extra code is emitted to convert between the two.
+  void visitCallAllAny(spv::Op OC, CallInst *CI);
+
+  /// Transform atomic_* to __spirv_Atomic*.
+  /// atomic_x(ptr_arg, args, order, scope) =>
+  ///   __spirv_AtomicY(ptr_arg, map(order), map(scope), args)
+  void transAtomicBuiltin(CallInst *CI, OCLBuiltinTransInfo &Info);
+
+  /// Transform atomic_work_item_fence to __spirv_MemoryBarrier.
+  /// atomic_work_item_fence(flag, order, scope) =>
+  ///   __spirv_MemoryBarrier(map(scope), map(flag)|map(order))
+  void visitCallAtomicWorkItemFence(CallInst *CI);
+
+  /// Transform atomic_compare_exchange call.
+  /// In atomic_compare_exchange, the expected value parameter is a pointer.
+  /// However in SPIR-V it is a value. The transformation adds a load
+  /// instruction, result of which is passed to atomic_compare_exchange as
+  /// argument.
+  /// The transformation adds a store instruction after the call, to update the
+  /// value in expected with the value pointed to by object. Though, it is not
+  /// necessary in case they are equal, this approach makes result code simpler.
+  /// Also ICmp instruction is added, because the call must return result of
+  /// comparison.
+  /// \returns the call instruction of atomic_compare_exchange_strong.
+  CallInst *visitCallAtomicCmpXchg(CallInst *CI);
+
+  /// Transform atomic_init.
+  /// atomic_init(p, x) => store p, x
+  void visitCallAtomicInit(CallInst *CI);
+
+  /// Transform legacy OCL 1.x atomic builtins to SPIR-V builtins for extensions
+  ///   cl_khr_int64_base_atomics
+  ///   cl_khr_int64_extended_atomics
+  /// Do nothing if the called function is not a legacy atomic builtin.
+  void visitCallAtomicLegacy(CallInst *CI, StringRef MangledName,
+                             StringRef DemangledName);
+
+  /// Transform OCL 2.0 C++11 atomic builtins to SPIR-V builtins.
+  /// Do nothing if the called function is not a C++11 atomic builtin.
+  void visitCallAtomicCpp11(CallInst *CI, StringRef MangledName,
+                            StringRef DemangledName);
+
+  /// Transform OCL builtin function to SPIR-V builtin function.
+  /// Assuming there is a simple name mapping without argument changes.
+  /// Should be called at last.
+  void visitCallBuiltinSimple(CallInst *CI, StringRef MangledName,
+                              StringRef DemangledName);
+
+  /// Transform get_image_{width|height|depth|dim}.
+  /// get_image_xxx(...) =>
+  ///   dimension = __spirv_ImageQuerySizeLod_R{ReturnType}(...);
+  ///   return dimension.{x|y|z};
+  void visitCallGetImageSize(CallInst *CI, StringRef DemangledName);
+
+  /// Transform {work|sub}_group_x =>
+  ///   __spirv_{OpName}
+  ///
+  /// Special handling of work_group_broadcast.
+  ///   work_group_broadcast(a, x, y, z)
+  ///     =>
+  ///   __spirv_GroupBroadcast(a, vec3(x, y, z))
+
+  void visitCallGroupBuiltin(CallInst *CI, StringRef DemangledName);
+
+  /// Transform mem_fence to __spirv_MemoryBarrier.
+  /// mem_fence(flag) => __spirv_MemoryBarrier(Workgroup, map(flag))
+  void visitCallMemFence(CallInst *CI, StringRef DemangledName);
+
+  void visitCallNDRange(CallInst *CI, StringRef DemangledName);
+
+  /// Transform read_image with sampler arguments.
+  /// read_image(image, sampler, ...) =>
+  ///   sampled_image = __spirv_SampledImage(image, sampler);
+  ///   return __spirv_ImageSampleExplicitLod_R{ReturnType}(sampled_image, ...);
+  void visitCallReadImageWithSampler(CallInst *CI, StringRef MangledName);
+
+  /// Transform read_image with msaa image arguments.
+  /// Sample argument must be acoded as Image Operand.
+  void visitCallReadImageMSAA(CallInst *CI, StringRef MangledName);
+
+  /// Transform {read|write}_image without sampler arguments.
+  void visitCallReadWriteImage(CallInst *CI, StringRef DemangledName);
+
+  /// Transform to_{global|local|private}.
+  ///
+  /// T* a = ...;
+  /// addr T* b = to_addr(a);
+  ///   =>
+  /// i8* x = cast<i8*>(a);
+  /// addr i8* y = __spirv_GenericCastToPtr_ToAddr(x);
+  /// addr T* b = cast<addr T*>(y);
+  void visitCallToAddr(CallInst *CI, StringRef DemangledName);
+
+  /// Transform return type of relatinal built-in functions like isnan, isfinite
+  /// to boolean values.
+  void visitCallRelational(CallInst *CI, StringRef DemangledName);
+
+  /// Transform vector load/store functions to SPIR-V extended builtin
+  ///   functions
+  /// {vload|vstore{a}}{_half}{n}{_rte|_rtz|_rtp|_rtn} =>
+  ///   __spirv_ocl_{ExtendedInstructionOpCodeName}__R{ReturnType}
+  void visitCallVecLoadStore(CallInst *CI, StringRef MangledName,
+                             StringRef DemangledName);
+
+  /// Transforms get_mem_fence built-in to SPIR-V function and aligns result
+  /// values with SPIR 1.2. get_mem_fence(ptr) => __spirv_GenericPtrMemSemantics
+  /// GenericPtrMemSemantics valid values are 0x100, 0x200 and 0x300, where is
+  /// SPIR 1.2 defines them as 0x1, 0x2 and 0x3, so this function adjusts
+  /// GenericPtrMemSemantics results to SPIR 1.2 values.
+  void visitCallGetFence(CallInst *CI, StringRef DemangledName);
+
+  /// Transforms OpDot instructions with a scalar type to a fmul instruction
+  void visitCallDot(CallInst *CI);
+
+  /// Fixes for built-in functions with vector+scalar arguments that are
+  /// translated to the SPIR-V instructions where all arguments must have the
+  /// same type.
+  void visitCallScalToVec(CallInst *CI, StringRef MangledName,
+                          StringRef DemangledName);
+
+  /// Transform get_image_channel_{order|data_type} built-in functions to
+  ///   __spirv_ocl_{ImageQueryOrder|ImageQueryFormat}
+  void visitCallGetImageChannel(CallInst *CI, StringRef DemangledName,
+                                unsigned int Offset);
+
+  /// Transform enqueue_kernel and kernel query built-in functions to
+  /// spirv-friendly format filling arguments, required for device-side enqueue
+  /// instructions, but missed in the original call
+  void visitCallEnqueueKernel(CallInst *CI, StringRef DemangledName);
+  void visitCallKernelQuery(CallInst *CI, StringRef DemangledName);
+
+  /// For cl_intel_subgroups block read built-ins:
+  void visitSubgroupBlockReadINTEL(CallInst *CI);
+
+  /// For cl_intel_subgroups block write built-ins:
+  void visitSubgroupBlockWriteINTEL(CallInst *CI);
+
+  /// For cl_intel_media_block_io built-ins:
+  void visitSubgroupImageMediaBlockINTEL(CallInst *CI, StringRef DemangledName);
+  // For cl_intel_device_side_avc_motion_estimation built-ins
+  void visitSubgroupAVCBuiltinCall(CallInst *CI, StringRef DemangledName);
+  void visitSubgroupAVCWrapperBuiltinCall(CallInst *CI, Op WrappedOC,
+                                          StringRef DemangledName);
+  void visitSubgroupAVCBuiltinCallWithSampler(CallInst *CI,
+                                              StringRef DemangledName);
+
+  void visitCallLdexp(CallInst *CI, StringRef MangledName,
+                      StringRef DemangledName);
+
+  /// For cl_intel_convert_bfloat16_as_ushort
+  void visitCallConvertBFloat16AsUshort(CallInst *CI, StringRef DemangledName);
+  /// For cl_intel_convert_as_bfloat16_float
+  void visitCallConvertAsBFloat16Float(CallInst *CI, StringRef DemangledName);
+
+  void setOCLTypeToSPIRV(OCLTypeToSPIRVBase *OCLTypeToSPIRV) {
+    OCLTypeToSPIRVPtr = OCLTypeToSPIRV;
+  }
+  OCLTypeToSPIRVBase *getOCLTypeToSPIRV() { return OCLTypeToSPIRVPtr; }
+
+private:
+  Module *M;
+  LLVMContext *Ctx;
+  unsigned CLVer; /// OpenCL version as major*10+minor
+  std::set<Value *> ValuesToDelete;
+  OCLTypeToSPIRVBase *OCLTypeToSPIRVPtr;
+
+  ConstantInt *addInt32(int I) { return getInt32(M, I); }
+  ConstantInt *addSizet(uint64_t I) { return getSizet(M, I); }
+
+  /// Get vector width from OpenCL vload* function name.
+  SPIRVWord getVecLoadWidth(const std::string &DemangledName);
+
+  /// Transform OpenCL vload/vstore function name.
+  void transVecLoadStoreName(std::string &DemangledName,
+                             const std::string &Stem, bool AlwaysN);
+};
+
+class OCLToSPIRVLegacy : public OCLToSPIRVBase, public llvm::ModulePass {
+public:
+  OCLToSPIRVLegacy() : ModulePass(ID) {
+    initializeOCLToSPIRVLegacyPass(*PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override;
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override;
+
+  static char ID;
+};
+
+class OCLToSPIRVPass : public OCLToSPIRVBase,
+                       public llvm::PassInfoMixin<OCLToSPIRVPass> {
+public:
+  llvm::PreservedAnalyses run(llvm::Module &M,
+                              llvm::ModuleAnalysisManager &MAM);
+};
+
+} // namespace SPIRV
+
+#endif // SPIRV_OCLTOSPIRV_H
diff --git a/lib/SPIRV/OCLTypeToSPIRV.cpp b/lib/SPIRV/OCLTypeToSPIRV.cpp
index 2721baaf..288a2ea3 100644
--- a/lib/SPIRV/OCLTypeToSPIRV.cpp
+++ b/lib/SPIRV/OCLTypeToSPIRV.cpp
@@ -327,6 +327,8 @@ Type *OCLTypeToSPIRVBase::getAdaptedType(Value *V) {
 
 } // namespace SPIRV
 
+AnalysisKey OCLTypeToSPIRVPass::Key;
+
 INITIALIZE_PASS(OCLTypeToSPIRVLegacy, "cltytospv", "Adapt OCL types for SPIR-V",
                 false, true)
 
diff --git a/lib/SPIRV/PreprocessMetadata.cpp b/lib/SPIRV/PreprocessMetadata.cpp
index a0a653fd..e59fcfb4 100644
--- a/lib/SPIRV/PreprocessMetadata.cpp
+++ b/lib/SPIRV/PreprocessMetadata.cpp
@@ -38,6 +38,7 @@
 //===----------------------------------------------------------------------===//
 #define DEBUG_TYPE "clmdtospv"
 
+#include "PreprocessMetadata.h"
 #include "OCLUtil.h"
 #include "SPIRVInternal.h"
 #include "SPIRVMDBuilder.h"
@@ -48,8 +49,6 @@
 #include "llvm/ADT/Triple.h"
 #include "llvm/IR/IRBuilder.h"
 #include "llvm/IR/InstVisitor.h"
-#include "llvm/IR/PassManager.h"
-#include "llvm/Pass.h"
 #include "llvm/Support/CommandLine.h"
 
 using namespace llvm;
@@ -61,42 +60,6 @@ namespace SPIRV {
 cl::opt<bool> EraseOCLMD("spirv-erase-cl-md", cl::init(true),
                          cl::desc("Erase OpenCL metadata"));
 
-class PreprocessMetadataBase {
-public:
-  PreprocessMetadataBase() : M(nullptr), Ctx(nullptr) {}
-
-  bool runPreprocessMetadata(Module &M);
-  void visit(Module *M);
-  void preprocessCXXStructorList(SPIRVMDBuilder::NamedMDWrapper &EM,
-                                 GlobalVariable *V, ExecutionMode EMode);
-  void preprocessOCLMetadata(Module *M, SPIRVMDBuilder *B, SPIRVMDWalker *W);
-  void preprocessVectorComputeMetadata(Module *M, SPIRVMDBuilder *B,
-                                       SPIRVMDWalker *W);
-
-private:
-  Module *M;
-  LLVMContext *Ctx;
-};
-
-class PreprocessMetadataLegacy : public ModulePass,
-                                 public PreprocessMetadataBase {
-public:
-  PreprocessMetadataLegacy() : ModulePass(ID) {
-    initializePreprocessMetadataLegacyPass(*PassRegistry::getPassRegistry());
-  }
-  bool runOnModule(Module &M) override;
-
-  static char ID;
-};
-
-class PreprocessMetadataPass
-    : public llvm::PassInfoMixin<PreprocessMetadataPass>,
-      public PreprocessMetadataBase {
-public:
-  llvm::PreservedAnalyses run(llvm::Module &M,
-                              llvm::ModuleAnalysisManager &MAM);
-};
-
 char PreprocessMetadataLegacy::ID = 0;
 
 bool PreprocessMetadataLegacy::runOnModule(Module &Module) {
diff --git a/lib/SPIRV/PreprocessMetadata.h b/lib/SPIRV/PreprocessMetadata.h
new file mode 100644
index 00000000..c4e85cdf
--- /dev/null
+++ b/lib/SPIRV/PreprocessMetadata.h
@@ -0,0 +1,90 @@
+//=- PreprocessMetadata.h - Metadata preprocessing pass -*- C++ -*-=//
+//
+//                     The LLVM/SPIR-V Translator
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright (c) 2022 The Khronos Group Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of The Khronos Group, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements preprocessing of LLVM IR metadata in order to perform
+// further translation to SPIR-V.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SPIRV_PREPROCESSMETADATA_H
+#define SPIRV_PREPROCESSMETADATA_H
+
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+#include "SPIRVMDBuilder.h"
+
+namespace SPIRV {
+
+class SPIRVMDWalker;
+
+class PreprocessMetadataBase {
+public:
+  PreprocessMetadataBase() : M(nullptr), Ctx(nullptr) {}
+
+  bool runPreprocessMetadata(Module &M);
+  void visit(Module *M);
+  void preprocessCXXStructorList(SPIRVMDBuilder::NamedMDWrapper &EM,
+                                 GlobalVariable *V, ExecutionMode EMode);
+  void preprocessOCLMetadata(Module *M, SPIRVMDBuilder *B, SPIRVMDWalker *W);
+  void preprocessVectorComputeMetadata(Module *M, SPIRVMDBuilder *B,
+                                       SPIRVMDWalker *W);
+
+private:
+  Module *M;
+  LLVMContext *Ctx;
+};
+
+class PreprocessMetadataLegacy : public ModulePass,
+                                 public PreprocessMetadataBase {
+public:
+  PreprocessMetadataLegacy() : ModulePass(ID) {
+    initializePreprocessMetadataLegacyPass(*PassRegistry::getPassRegistry());
+  }
+  bool runOnModule(Module &M) override;
+
+  static char ID;
+};
+
+class PreprocessMetadataPass
+    : public llvm::PassInfoMixin<PreprocessMetadataPass>,
+      public PreprocessMetadataBase {
+public:
+  llvm::PreservedAnalyses run(llvm::Module &M,
+                              llvm::ModuleAnalysisManager &MAM);
+};
+
+} // namespace SPIRV
+
+#endif // SPIRV_PREPROCESSMETADATA_H
diff --git a/lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp b/lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp
index d7a0e9ca..900cf45e 100644
--- a/lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp
+++ b/lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp
@@ -40,12 +40,12 @@
 // point types, 2/3/4/8/16-element vector of scalar types").
 //
 //===----------------------------------------------------------------------===//
+
+#include "SPIRVLowerBitCastToNonStandardType.h"
 #include "SPIRVInternal.h"
 
 #include "llvm/IR/IRBuilder.h"
 #include "llvm/IR/NoFolder.h"
-#include "llvm/IR/PassManager.h"
-#include "llvm/Pass.h"
 #include "llvm/Transforms/Utils/Local.h"
 
 #include <utility>
@@ -119,135 +119,116 @@ static bool isNonStdVecType(VectorType *VecTy) {
   return !isValidVectorSize(NumElems);
 }
 
-class SPIRVLowerBitCastToNonStandardTypePass
-    : public llvm::PassInfoMixin<SPIRVLowerBitCastToNonStandardTypePass> {
-public:
-  SPIRVLowerBitCastToNonStandardTypePass(const SPIRV::TranslatorOpts &Opts)
-      : Opts(Opts) {}
-
-  PreservedAnalyses
-  runLowerBitCastToNonStandardType(Function &F, FunctionAnalysisManager &FAM) {
-    // This pass doesn't cover all possible uses of non-standard types, only
-    // known. We assume that bad type won't be passed to a function as
-    // parameter, since it added by an optimization.
-    bool Changed = false;
-
-    // SPV_INTEL_vector_compute allows to use vectors with any number of
-    // components. Since this method only lowers vectors with non-standard
-    // in pure SPIR-V number of components, there is no need to do anything in
-    // case SPV_INTEL_vector_compute is enabled.
-    if (Opts.isAllowedToUseExtension(ExtensionID::SPV_INTEL_vector_compute))
-      return PreservedAnalyses::all();
-
-    // The basic pattern we're trying to fix is this InstCombine pattern:
-    // trunc (extractelement) -> extractelement (bitcast)
-    // (note that the bitcast itself can get propagated back to change the type
-    // of load instructions, and even through those to pointer casts, if typed
-    // pointers are enabled.
-    std::vector<ExtractElementInst *> NonStdVecInsts;
-    SmallVector<WeakTrackingVH, 4> MaybeDeletedInsts;
-    for (auto &BB : F)
-      for (auto &I : BB) {
-        if (auto *EI = dyn_cast<ExtractElementInst>(&I)) {
-          if (isNonStdVecType(EI->getVectorOperandType()))
-            NonStdVecInsts.push_back(EI);
-        } else if (auto *VT = dyn_cast<VectorType>(I.getType())) {
-          if (isNonStdVecType(VT)) {
-            MaybeDeletedInsts.push_back(&I);
-          }
+PreservedAnalyses
+SPIRVLowerBitCastToNonStandardTypePass::run(Function &F,
+                                            FunctionAnalysisManager &FAM) {
+  // This pass doesn't cover all possible uses of non-standard types, only
+  // known. We assume that bad type won't be passed to a function as
+  // parameter, since it added by an optimization.
+  bool Changed = false;
+
+  // SPV_INTEL_vector_compute allows to use vectors with any number of
+  // components. Since this method only lowers vectors with non-standard
+  // in pure SPIR-V number of components, there is no need to do anything in
+  // case SPV_INTEL_vector_compute is enabled.
+  if (Opts.isAllowedToUseExtension(ExtensionID::SPV_INTEL_vector_compute))
+    return PreservedAnalyses::all();
+
+  // The basic pattern we're trying to fix is this InstCombine pattern:
+  // trunc (extractelement) -> extractelement (bitcast)
+  // (note that the bitcast itself can get propagated back to change the type
+  // of load instructions, and even through those to pointer casts, if typed
+  // pointers are enabled.
+  std::vector<ExtractElementInst *> NonStdVecInsts;
+  SmallVector<WeakTrackingVH, 4> MaybeDeletedInsts;
+  for (auto &BB : F)
+    for (auto &I : BB) {
+      if (auto *EI = dyn_cast<ExtractElementInst>(&I)) {
+        if (isNonStdVecType(EI->getVectorOperandType()))
+          NonStdVecInsts.push_back(EI);
+      } else if (auto *VT = dyn_cast<VectorType>(I.getType())) {
+        if (isNonStdVecType(VT)) {
+          MaybeDeletedInsts.push_back(&I);
         }
       }
-
-    std::vector<Instruction *> InstsToErase;
-    NFIRBuilder Builder(F.getContext());
-    for (auto &I : NonStdVecInsts) {
-      VectorType *OldVecTy = I->getVectorOperandType();
-      unsigned OldVecSize = OldVecTy->getElementCount().getFixedValue();
-
-      // Compute the adjustment factor for the new vector size.
-      unsigned VecFactor = 2;
-      while (OldVecSize % VecFactor == 0 &&
-             !isValidVectorSize(OldVecSize / VecFactor))
-        VecFactor *= 2;
-      if (OldVecSize % VecFactor != 0) {
-        report_fatal_error(Twine("Invalid vector size for fixup: ") +
-                           Twine(OldVecSize));
-        return PreservedAnalyses::none();
-      }
-      unsigned NewElemSize = OldVecTy->getScalarSizeInBits() * VecFactor;
-      VectorType *NewVecTy =
-          VectorType::get(Type::getIntNTy(F.getContext(), NewElemSize),
-                          OldVecSize / VecFactor, false);
-
-      // Adjust the element index as appropriate.
-      uint64_t OldElemIdx =
-          cast<ConstantInt>(I->getIndexOperand())->getZExtValue();
-      uint64_t NewElemIdx = OldElemIdx / VecFactor;
-      uint64_t ShiftCount = OldElemIdx % VecFactor;
-      Builder.SetInsertPoint(I);
-      Value *NewVecOp = removeBitCasts(I->getVectorOperand(), NewVecTy, Builder,
-                                       InstsToErase);
-      Value *NewExtracted = Builder.CreateExtractElement(NewVecOp, NewElemIdx);
-
-      // If the extract does higher-order bits of the value, shift as necessary.
-      if (ShiftCount > 0)
-        NewExtracted = Builder.CreateLShr(
-            NewExtracted, ShiftCount * OldVecTy->getScalarSizeInBits());
-
-      Value *NewValue = Builder.CreateTrunc(NewExtracted, I->getType());
-      I->replaceAllUsesWith(NewValue);
-      I->eraseFromParent();
-      Changed = true;
     }
 
-    for (auto *I : InstsToErase)
-      RecursivelyDeleteTriviallyDeadInstructions(I);
-
-    // Check if there are any residual unsupported vector types.
-    for (auto &VH : MaybeDeletedInsts) {
-      // Some vector-valued instructions were replaced with undef values, so if
-      // that's what we got, it's still a dead instruction.
-      if (VH.pointsToAliveValue() && !isa<UndefValue>(VH)) {
-        auto *VT = dyn_cast<VectorType>(VH->getType());
-        report_fatal_error(Twine("Unsupported vector type with ") +
-                               Twine(VT->getElementCount().getFixedValue()) +
-                               Twine(" elements"),
-                           false);
-      }
+  std::vector<Instruction *> InstsToErase;
+  NFIRBuilder Builder(F.getContext());
+  for (auto &I : NonStdVecInsts) {
+    VectorType *OldVecTy = I->getVectorOperandType();
+    unsigned OldVecSize = OldVecTy->getElementCount().getFixedValue();
+
+    // Compute the adjustment factor for the new vector size.
+    unsigned VecFactor = 2;
+    while (OldVecSize % VecFactor == 0 &&
+           !isValidVectorSize(OldVecSize / VecFactor))
+      VecFactor *= 2;
+    if (OldVecSize % VecFactor != 0) {
+      report_fatal_error(Twine("Invalid vector size for fixup: ") +
+                         Twine(OldVecSize));
+      return PreservedAnalyses::none();
     }
-
-    return Changed ? PreservedAnalyses::none() : PreservedAnalyses::all();
+    unsigned NewElemSize = OldVecTy->getScalarSizeInBits() * VecFactor;
+    VectorType *NewVecTy =
+        VectorType::get(Type::getIntNTy(F.getContext(), NewElemSize),
+                        OldVecSize / VecFactor, false);
+
+    // Adjust the element index as appropriate.
+    uint64_t OldElemIdx =
+        cast<ConstantInt>(I->getIndexOperand())->getZExtValue();
+    uint64_t NewElemIdx = OldElemIdx / VecFactor;
+    uint64_t ShiftCount = OldElemIdx % VecFactor;
+    Builder.SetInsertPoint(I);
+    Value *NewVecOp =
+        removeBitCasts(I->getVectorOperand(), NewVecTy, Builder, InstsToErase);
+    Value *NewExtracted = Builder.CreateExtractElement(NewVecOp, NewElemIdx);
+
+    // If the extract does higher-order bits of the value, shift as necessary.
+    if (ShiftCount > 0)
+      NewExtracted = Builder.CreateLShr(
+          NewExtracted, ShiftCount * OldVecTy->getScalarSizeInBits());
+
+    Value *NewValue = Builder.CreateTrunc(NewExtracted, I->getType());
+    I->replaceAllUsesWith(NewValue);
+    I->eraseFromParent();
+    Changed = true;
   }
 
-private:
-  SPIRV::TranslatorOpts Opts;
-};
-
-class SPIRVLowerBitCastToNonStandardTypeLegacy : public FunctionPass {
-public:
-  static char ID;
-  SPIRVLowerBitCastToNonStandardTypeLegacy(const SPIRV::TranslatorOpts &Opts)
-      : FunctionPass(ID), Opts(Opts) {}
-
-  SPIRVLowerBitCastToNonStandardTypeLegacy() : FunctionPass(ID) {}
-
-  bool runOnFunction(Function &F) override {
-    SPIRVLowerBitCastToNonStandardTypePass Impl(Opts);
-    FunctionAnalysisManager FAM;
-    auto PA = Impl.runLowerBitCastToNonStandardType(F, FAM);
-    return !PA.areAllPreserved();
+  for (auto *I : InstsToErase)
+    RecursivelyDeleteTriviallyDeadInstructions(I);
+
+  // Check if there are any residual unsupported vector types.
+  for (auto &VH : MaybeDeletedInsts) {
+    // Some vector-valued instructions were replaced with undef values, so if
+    // that's what we got, it's still a dead instruction.
+    if (VH.pointsToAliveValue() && !isa<UndefValue>(VH)) {
+      auto *VT = dyn_cast<VectorType>(VH->getType());
+      report_fatal_error(Twine("Unsupported vector type with ") +
+                             Twine(VT->getElementCount().getFixedValue()) +
+                             Twine(" elements"),
+                         false);
+    }
   }
 
-  bool doFinalization(Module &M) override {
-    verifyRegularizationPass(M, "SPIRVLowerBitCastToNonStandardType");
-    return false;
-  }
+  return Changed ? PreservedAnalyses::none() : PreservedAnalyses::all();
+}
 
-  StringRef getPassName() const override { return "Lower nonstandard type"; }
+bool SPIRVLowerBitCastToNonStandardTypeLegacy::runOnFunction(Function &F) {
+  SPIRVLowerBitCastToNonStandardTypePass Impl(Opts);
+  FunctionAnalysisManager FAM;
+  auto PA = Impl.run(F, FAM);
+  return !PA.areAllPreserved();
+}
 
-private:
-  SPIRV::TranslatorOpts Opts;
-};
+bool SPIRVLowerBitCastToNonStandardTypeLegacy::doFinalization(Module &M) {
+  verifyRegularizationPass(M, "SPIRVLowerBitCastToNonStandardType");
+  return false;
+}
+
+StringRef SPIRVLowerBitCastToNonStandardTypeLegacy::getPassName() const {
+  return "Lower nonstandard type";
+}
 
 char SPIRVLowerBitCastToNonStandardTypeLegacy::ID = 0;
 
diff --git a/lib/SPIRV/SPIRVLowerBitCastToNonStandardType.h b/lib/SPIRV/SPIRVLowerBitCastToNonStandardType.h
new file mode 100644
index 00000000..8a04d2a1
--- /dev/null
+++ b/lib/SPIRV/SPIRVLowerBitCastToNonStandardType.h
@@ -0,0 +1,79 @@
+//===- SPIRVLowerBitCastToNonStandardType.h - Bitcast lowering --*- C++ -*-===//
+//
+//                     The LLVM/SPIR-V Translator
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright (c) 2022 The Khronos Group Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of The Khronos Group, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SPIRV_SPIRVLOWERBITCASTTONONSTANDARDTYPE_H
+#define SPIRV_SPIRVLOWERBITCASTTONONSTANDARDTYPE_H
+
+#include "LLVMSPIRVOpts.h"
+
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+namespace SPIRV {
+
+class SPIRVLowerBitCastToNonStandardTypePass
+    : public llvm::PassInfoMixin<SPIRVLowerBitCastToNonStandardTypePass> {
+public:
+  SPIRVLowerBitCastToNonStandardTypePass(const SPIRV::TranslatorOpts &Opts)
+      : Opts(Opts) {}
+
+  PreservedAnalyses run(Function &F, FunctionAnalysisManager &FAM);
+
+private:
+  SPIRV::TranslatorOpts Opts;
+};
+
+class SPIRVLowerBitCastToNonStandardTypeLegacy : public FunctionPass {
+public:
+  static char ID;
+  SPIRVLowerBitCastToNonStandardTypeLegacy(const SPIRV::TranslatorOpts &Opts)
+      : FunctionPass(ID), Opts(Opts) {}
+
+  SPIRVLowerBitCastToNonStandardTypeLegacy() : FunctionPass(ID) {}
+
+  bool runOnFunction(Function &F) override;
+
+  bool doFinalization(Module &M) override;
+
+  StringRef getPassName() const override;
+
+private:
+  SPIRV::TranslatorOpts Opts;
+};
+
+} // namespace SPIRV
+
+#endif // SPIRV_SPIRVLOWERBITCASTTONONSTANDARDTYPE_H
diff --git a/lib/SPIRV/SPIRVLowerBool.cpp b/lib/SPIRV/SPIRVLowerBool.cpp
index 9b1f3bbf..f9ef6d8f 100644
--- a/lib/SPIRV/SPIRVLowerBool.cpp
+++ b/lib/SPIRV/SPIRVLowerBool.cpp
@@ -37,112 +37,110 @@
 //===----------------------------------------------------------------------===//
 #define DEBUG_TYPE "spvbool"
 
+#include "SPIRVLowerBool.h"
 #include "SPIRVInternal.h"
 #include "libSPIRV/SPIRVDebug.h"
 
 #include "llvm/IR/IRBuilder.h"
-#include "llvm/IR/InstVisitor.h"
-#include "llvm/IR/Instructions.h"
-#include "llvm/IR/PassManager.h"
-#include "llvm/Pass.h"
 
 using namespace llvm;
 using namespace SPIRV;
 
 namespace SPIRV {
 
-class SPIRVLowerBoolBase : public InstVisitor<SPIRVLowerBoolBase> {
-public:
-  SPIRVLowerBoolBase() : Context(nullptr) {}
-  virtual ~SPIRVLowerBoolBase() {}
-  void replace(Instruction *I, Instruction *NewI) {
-    NewI->takeName(I);
-    NewI->setDebugLoc(I->getDebugLoc());
-    I->replaceAllUsesWith(NewI);
-    I->dropAllReferences();
-    I->eraseFromParent();
-  }
-  bool isBoolType(Type *Ty) {
-    if (Ty->isIntegerTy(1))
-      return true;
-    if (auto VT = dyn_cast<VectorType>(Ty))
-      return isBoolType(VT->getElementType());
-    return false;
-  }
-  virtual void visitTruncInst(TruncInst &I) {
-    if (isBoolType(I.getType())) {
-      auto Op = I.getOperand(0);
-      auto And = BinaryOperator::CreateAnd(
-          Op, getScalarOrVectorConstantInt(Op->getType(), 1, false), "", &I);
-      And->setDebugLoc(I.getDebugLoc());
-      auto Zero = getScalarOrVectorConstantInt(Op->getType(), 0, false);
-      auto Cmp = new ICmpInst(&I, CmpInst::ICMP_NE, And, Zero);
-      replace(&I, Cmp);
-    }
-  }
-  void handleExtInstructions(Instruction &I) {
-    auto Op = I.getOperand(0);
-    if (isBoolType(Op->getType())) {
-      auto Opcode = I.getOpcode();
-      auto Ty = I.getType();
-      auto Zero = getScalarOrVectorConstantInt(Ty, 0, false);
-      auto One = getScalarOrVectorConstantInt(
-          Ty, (Opcode == Instruction::SExt) ? ~0 : 1, false);
-      assert(Zero && One && "Couldn't create constant int");
-      auto Sel = SelectInst::Create(Op, One, Zero, "", &I);
-      replace(&I, Sel);
-    }
-  }
-  void handleCastInstructions(Instruction &I) {
-    auto Op = I.getOperand(0);
-    auto *OpTy = Op->getType();
-    if (isBoolType(OpTy)) {
-      Type *Ty = Type::getInt32Ty(*Context);
-      if (auto VT = dyn_cast<FixedVectorType>(OpTy))
-        Ty = llvm::FixedVectorType::get(Ty, VT->getNumElements());
-      auto Zero = getScalarOrVectorConstantInt(Ty, 0, false);
-      auto One = getScalarOrVectorConstantInt(Ty, 1, false);
-      assert(Zero && One && "Couldn't create constant int");
-      auto Sel = SelectInst::Create(Op, One, Zero, "", &I);
-      Sel->setDebugLoc(I.getDebugLoc());
-      I.setOperand(0, Sel);
-    }
-  }
-  virtual void visitZExtInst(ZExtInst &I) { handleExtInstructions(I); }
-  virtual void visitSExtInst(SExtInst &I) { handleExtInstructions(I); }
-  virtual void visitUIToFPInst(UIToFPInst &I) { handleCastInstructions(I); }
-  virtual void visitSIToFPInst(SIToFPInst &I) { handleCastInstructions(I); }
-  bool runLowerBool(Module &M) {
-    Context = &M.getContext();
-    visit(M);
-
-    verifyRegularizationPass(M, "SPIRVLowerBool");
+void SPIRVLowerBoolBase::replace(Instruction *I, Instruction *NewI) {
+  NewI->takeName(I);
+  NewI->setDebugLoc(I->getDebugLoc());
+  I->replaceAllUsesWith(NewI);
+  I->dropAllReferences();
+  I->eraseFromParent();
+}
+
+bool SPIRVLowerBoolBase::isBoolType(Type *Ty) {
+  if (Ty->isIntegerTy(1))
     return true;
+  if (auto VT = dyn_cast<VectorType>(Ty))
+    return isBoolType(VT->getElementType());
+  return false;
+}
+
+void SPIRVLowerBoolBase::visitTruncInst(TruncInst &I) {
+  if (isBoolType(I.getType())) {
+    auto Op = I.getOperand(0);
+    auto And = BinaryOperator::CreateAnd(
+        Op, getScalarOrVectorConstantInt(Op->getType(), 1, false), "", &I);
+    And->setDebugLoc(I.getDebugLoc());
+    auto Zero = getScalarOrVectorConstantInt(Op->getType(), 0, false);
+    auto Cmp = new ICmpInst(&I, CmpInst::ICMP_NE, And, Zero);
+    replace(&I, Cmp);
   }
+}
 
-private:
-  LLVMContext *Context;
-};
-
-class SPIRVLowerBoolPass : public llvm::PassInfoMixin<SPIRVLowerBoolPass>,
-                           public SPIRVLowerBoolBase {
-public:
-  llvm::PreservedAnalyses run(llvm::Module &M,
-                              llvm::ModuleAnalysisManager &MAM) {
-    return runLowerBool(M) ? llvm::PreservedAnalyses::none()
-                           : llvm::PreservedAnalyses::all();
+void SPIRVLowerBoolBase::handleExtInstructions(Instruction &I) {
+  auto Op = I.getOperand(0);
+  if (isBoolType(Op->getType())) {
+    auto Opcode = I.getOpcode();
+    auto Ty = I.getType();
+    auto Zero = getScalarOrVectorConstantInt(Ty, 0, false);
+    auto One = getScalarOrVectorConstantInt(
+        Ty, (Opcode == Instruction::SExt) ? ~0 : 1, false);
+    assert(Zero && One && "Couldn't create constant int");
+    auto Sel = SelectInst::Create(Op, One, Zero, "", &I);
+    replace(&I, Sel);
   }
-};
+}
 
-class SPIRVLowerBoolLegacy : public ModulePass, public SPIRVLowerBoolBase {
-public:
-  SPIRVLowerBoolLegacy() : ModulePass(ID) {
-    initializeSPIRVLowerBoolLegacyPass(*PassRegistry::getPassRegistry());
+void SPIRVLowerBoolBase::handleCastInstructions(Instruction &I) {
+  auto Op = I.getOperand(0);
+  auto *OpTy = Op->getType();
+  if (isBoolType(OpTy)) {
+    Type *Ty = Type::getInt32Ty(*Context);
+    if (auto VT = dyn_cast<FixedVectorType>(OpTy))
+      Ty = llvm::FixedVectorType::get(Ty, VT->getNumElements());
+    auto Zero = getScalarOrVectorConstantInt(Ty, 0, false);
+    auto One = getScalarOrVectorConstantInt(Ty, 1, false);
+    assert(Zero && One && "Couldn't create constant int");
+    auto Sel = SelectInst::Create(Op, One, Zero, "", &I);
+    Sel->setDebugLoc(I.getDebugLoc());
+    I.setOperand(0, Sel);
   }
-  bool runOnModule(Module &M) override { return runLowerBool(M); }
+}
+
+void SPIRVLowerBoolBase::visitZExtInst(ZExtInst &I) {
+  handleExtInstructions(I);
+}
+
+void SPIRVLowerBoolBase::visitSExtInst(SExtInst &I) {
+  handleExtInstructions(I);
+}
+
+void SPIRVLowerBoolBase::visitUIToFPInst(UIToFPInst &I) {
+  handleCastInstructions(I);
+}
+
+void SPIRVLowerBoolBase::visitSIToFPInst(SIToFPInst &I) {
+  handleCastInstructions(I);
+}
+
+bool SPIRVLowerBoolBase::runLowerBool(Module &M) {
+  Context = &M.getContext();
+  visit(M);
+
+  verifyRegularizationPass(M, "SPIRVLowerBool");
+  return true;
+}
+
+llvm::PreservedAnalyses
+SPIRVLowerBoolPass::run(llvm::Module &M, llvm::ModuleAnalysisManager &MAM) {
+  return runLowerBool(M) ? llvm::PreservedAnalyses::none()
+                         : llvm::PreservedAnalyses::all();
+}
+
+SPIRVLowerBoolLegacy::SPIRVLowerBoolLegacy() : ModulePass(ID) {
+  initializeSPIRVLowerBoolLegacyPass(*PassRegistry::getPassRegistry());
+}
 
-  static char ID;
-};
+bool SPIRVLowerBoolLegacy::runOnModule(Module &M) { return runLowerBool(M); }
 
 char SPIRVLowerBoolLegacy::ID = 0;
 } // namespace SPIRV
diff --git a/lib/SPIRV/SPIRVLowerBool.h b/lib/SPIRV/SPIRVLowerBool.h
new file mode 100644
index 00000000..9c2f9851
--- /dev/null
+++ b/lib/SPIRV/SPIRVLowerBool.h
@@ -0,0 +1,87 @@
+//===- SPIRVLowerBool.h - Bool operand lowering  --------*- C++ -*-===//
+//
+//                     The LLVM/SPIR-V Translator
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright (c) 2022 The Khronos Group Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of The Khronos Group, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering instructions with bool operands.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SPIRV_SPIRVLOWERBOOL_H
+#define SPIRV_SPIRVLOWERBOOL_H
+
+#include "llvm/IR/InstVisitor.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+namespace SPIRV {
+
+class SPIRVLowerBoolBase : public InstVisitor<SPIRVLowerBoolBase> {
+public:
+  SPIRVLowerBoolBase() : Context(nullptr) {}
+  virtual ~SPIRVLowerBoolBase() {}
+  void replace(Instruction *I, Instruction *NewI);
+  bool isBoolType(Type *Ty);
+  virtual void visitTruncInst(TruncInst &I);
+  void handleExtInstructions(Instruction &I);
+  void handleCastInstructions(Instruction &I);
+  virtual void visitZExtInst(ZExtInst &I);
+  virtual void visitSExtInst(SExtInst &I);
+  virtual void visitUIToFPInst(UIToFPInst &I);
+  virtual void visitSIToFPInst(SIToFPInst &I);
+  bool runLowerBool(Module &M);
+
+private:
+  LLVMContext *Context;
+};
+
+class SPIRVLowerBoolPass : public llvm::PassInfoMixin<SPIRVLowerBoolPass>,
+                           public SPIRVLowerBoolBase {
+public:
+  llvm::PreservedAnalyses run(llvm::Module &M,
+                              llvm::ModuleAnalysisManager &MAM);
+};
+
+class SPIRVLowerBoolLegacy : public ModulePass, public SPIRVLowerBoolBase {
+public:
+  SPIRVLowerBoolLegacy();
+  bool runOnModule(Module &M) override;
+
+  static char ID;
+};
+
+} // namespace SPIRV
+
+#endif // SPIRV_SPIRVLOWERBOOL_H
diff --git a/lib/SPIRV/SPIRVLowerMemmove.cpp b/lib/SPIRV/SPIRVLowerMemmove.cpp
index 6bfdb701..a3f778d4 100644
--- a/lib/SPIRV/SPIRVLowerMemmove.cpp
+++ b/lib/SPIRV/SPIRVLowerMemmove.cpp
@@ -37,14 +37,11 @@
 //===----------------------------------------------------------------------===//
 #define DEBUG_TYPE "spvmemmove"
 
+#include "SPIRVLowerMemmove.h"
 #include "SPIRVInternal.h"
 #include "libSPIRV/SPIRVDebug.h"
 
 #include "llvm/IR/IRBuilder.h"
-#include "llvm/IR/IntrinsicInst.h"
-#include "llvm/IR/Module.h"
-#include "llvm/IR/PassManager.h"
-#include "llvm/Pass.h"
 #include "llvm/Transforms/Utils/LowerMemIntrinsics.h"
 
 using namespace llvm;
@@ -52,113 +49,100 @@ using namespace SPIRV;
 
 namespace SPIRV {
 
-class SPIRVLowerMemmoveBase {
-public:
-  SPIRVLowerMemmoveBase() : Context(nullptr) {}
-
-  void LowerMemMoveInst(MemMoveInst &I) {
-    // There is no direct equivalent of @llvm.memmove in SPIR-V and the closest
-    // instructions are 'OpCopyMemory' and 'OpCopyMemorySized'.
-    //
-    // 'OpCopyMemory' does not accept amount of bytes to copy and infers that
-    // from type which is being copied; also it only allows to copy value of a
-    // particular type to pointer pointing to the same type.
-    //
-    // 'OpCopyMemorySized' is closer to @llvm.memmove, because it actually
-    // copies bytes, but unlike memove it is not explicitly specified whether it
-    // supports overlapping source and destination. Therefore, we replace
-    // memmove with two 'OpCopyMemorySized' instructions: the first one copies
-    // bytes from source to a temporary location, the second one copies bytes
-    // from that temporary location to the destination.
-    IRBuilder<> Builder(I.getParent());
-    Builder.SetInsertPoint(&I);
-
-    auto *Length = cast<ConstantInt>(I.getLength());
-    auto *AllocaTy = ArrayType::get(IntegerType::getInt8Ty(*Context),
-                                    Length->getZExtValue());
-    MaybeAlign SrcAlign = I.getSourceAlign();
-
-    auto *Alloca = Builder.CreateAlloca(AllocaTy);
-    if (SrcAlign.hasValue())
-      Alloca->setAlignment(SrcAlign.getValue());
-
-    // FIXME: Do we need to pass the size of alloca here? From LangRef:
-    // > The first argument is a constant integer representing the size of the
-    // > object, or -1 if it is variable sized.
-    //
-    // https://llvm.org/docs/LangRef.html#llvm-lifetime-start-intrinsic
-    Builder.CreateLifetimeStart(Alloca);
-    Builder.CreateMemCpy(Alloca, SrcAlign, I.getRawSource(), SrcAlign, Length,
-                         I.isVolatile());
-
-    auto *SecondCpy =
-        Builder.CreateMemCpy(I.getRawDest(), I.getDestAlign(), Alloca, SrcAlign,
-                             Length, I.isVolatile());
-    Builder.CreateLifetimeEnd(Alloca);
-
-    SecondCpy->takeName(&I);
-    I.replaceAllUsesWith(SecondCpy);
-    I.dropAllReferences();
-    I.eraseFromParent();
-  }
+void SPIRVLowerMemmoveBase::LowerMemMoveInst(MemMoveInst &I) {
+  // There is no direct equivalent of @llvm.memmove in SPIR-V and the closest
+  // instructions are 'OpCopyMemory' and 'OpCopyMemorySized'.
+  //
+  // 'OpCopyMemory' does not accept amount of bytes to copy and infers that
+  // from type which is being copied; also it only allows to copy value of a
+  // particular type to pointer pointing to the same type.
+  //
+  // 'OpCopyMemorySized' is closer to @llvm.memmove, because it actually
+  // copies bytes, but unlike memove it is not explicitly specified whether it
+  // supports overlapping source and destination. Therefore, we replace
+  // memmove with two 'OpCopyMemorySized' instructions: the first one copies
+  // bytes from source to a temporary location, the second one copies bytes
+  // from that temporary location to the destination.
+  IRBuilder<> Builder(I.getParent());
+  Builder.SetInsertPoint(&I);
+
+  auto *Length = cast<ConstantInt>(I.getLength());
+  auto *AllocaTy =
+      ArrayType::get(IntegerType::getInt8Ty(*Context), Length->getZExtValue());
+  MaybeAlign SrcAlign = I.getSourceAlign();
+
+  auto *Alloca = Builder.CreateAlloca(AllocaTy);
+  if (SrcAlign.hasValue())
+    Alloca->setAlignment(SrcAlign.getValue());
+
+  // FIXME: Do we need to pass the size of alloca here? From LangRef:
+  // > The first argument is a constant integer representing the size of the
+  // > object, or -1 if it is variable sized.
+  //
+  // https://llvm.org/docs/LangRef.html#llvm-lifetime-start-intrinsic
+  Builder.CreateLifetimeStart(Alloca);
+  Builder.CreateMemCpy(Alloca, SrcAlign, I.getRawSource(), SrcAlign, Length,
+                       I.isVolatile());
+
+  auto *SecondCpy =
+      Builder.CreateMemCpy(I.getRawDest(), I.getDestAlign(), Alloca, SrcAlign,
+                           Length, I.isVolatile());
+  Builder.CreateLifetimeEnd(Alloca);
+
+  SecondCpy->takeName(&I);
+  I.replaceAllUsesWith(SecondCpy);
+  I.dropAllReferences();
+  I.eraseFromParent();
+}
 
-  bool expandMemMoveIntrinsicUses(Function &F) {
-    bool Changed = false;
-
-    for (User *U : make_early_inc_range(F.users())) {
-      MemMoveInst *Inst = cast<MemMoveInst>(U);
-      if (!isa<ConstantInt>(Inst->getLength())) {
-        expandMemMoveAsLoop(Inst);
-        Inst->eraseFromParent();
-      } else {
-        LowerMemMoveInst(*Inst);
-      }
-      Changed = true;
+bool SPIRVLowerMemmoveBase::expandMemMoveIntrinsicUses(Function &F) {
+  bool Changed = false;
+
+  for (User *U : make_early_inc_range(F.users())) {
+    MemMoveInst *Inst = cast<MemMoveInst>(U);
+    if (!isa<ConstantInt>(Inst->getLength())) {
+      expandMemMoveAsLoop(Inst);
+      Inst->eraseFromParent();
+    } else {
+      LowerMemMoveInst(*Inst);
     }
-    return Changed;
+    Changed = true;
   }
-  bool runLowerMemmove(Module &M) {
-    Context = &M.getContext();
-    bool Changed = false;
+  return Changed;
+}
 
-    for (Function &F : M) {
-      if (!F.isDeclaration())
-        continue;
+bool SPIRVLowerMemmoveBase::runLowerMemmove(Module &M) {
+  Context = &M.getContext();
+  bool Changed = false;
 
-      if (F.getIntrinsicID() == Intrinsic::memmove)
-        Changed |= expandMemMoveIntrinsicUses(F);
-    }
+  for (Function &F : M) {
+    if (!F.isDeclaration())
+      continue;
 
-    verifyRegularizationPass(M, "SPIRVLowerMemmove");
-    return Changed;
+    if (F.getIntrinsicID() == Intrinsic::memmove)
+      Changed |= expandMemMoveIntrinsicUses(F);
   }
 
-private:
-  LLVMContext *Context;
-};
-
-class SPIRVLowerMemmovePass : public llvm::PassInfoMixin<SPIRVLowerMemmovePass>,
-                              public SPIRVLowerMemmoveBase {
-public:
-  llvm::PreservedAnalyses run(llvm::Module &M,
-                              llvm::ModuleAnalysisManager &MAM) {
-    return runLowerMemmove(M) ? llvm::PreservedAnalyses::none()
-                              : llvm::PreservedAnalyses::all();
-  }
-};
+  verifyRegularizationPass(M, "SPIRVLowerMemmove");
+  return Changed;
+}
 
-class SPIRVLowerMemmoveLegacy : public ModulePass,
-                                public SPIRVLowerMemmoveBase {
-public:
-  SPIRVLowerMemmoveLegacy() : ModulePass(ID) {
-    initializeSPIRVLowerMemmoveLegacyPass(*PassRegistry::getPassRegistry());
-  }
-  bool runOnModule(Module &M) override { return runLowerMemmove(M); }
+llvm::PreservedAnalyses
+SPIRVLowerMemmovePass::run(llvm::Module &M, llvm::ModuleAnalysisManager &MAM) {
+  return runLowerMemmove(M) ? llvm::PreservedAnalyses::none()
+                            : llvm::PreservedAnalyses::all();
+}
+
+SPIRVLowerMemmoveLegacy::SPIRVLowerMemmoveLegacy() : ModulePass(ID) {
+  initializeSPIRVLowerMemmoveLegacyPass(*PassRegistry::getPassRegistry());
+}
 
-  static char ID;
-};
+bool SPIRVLowerMemmoveLegacy::runOnModule(Module &M) {
+  return runLowerMemmove(M);
+}
 
 char SPIRVLowerMemmoveLegacy::ID = 0;
+
 } // namespace SPIRV
 
 INITIALIZE_PASS(SPIRVLowerMemmoveLegacy, "spvmemmove",
diff --git a/lib/SPIRV/SPIRVLowerMemmove.h b/lib/SPIRV/SPIRVLowerMemmove.h
new file mode 100644
index 00000000..2647827d
--- /dev/null
+++ b/lib/SPIRV/SPIRVLowerMemmove.h
@@ -0,0 +1,81 @@
+//===- SPIRVLowerMemmove.h - memmove lowering  --------*- C++ -*-===//
+//
+//                     The LLVM/SPIR-V Translator
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright (c) 2022 The Khronos Group Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of The Khronos Group, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering llvm.memmove into several llvm.memcpys.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SPIRV_SPIRVLOWERMEMMOVE_H
+#define SPIRV_SPIRVLOWERMEMMOVE_H
+
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+namespace SPIRV {
+
+class SPIRVLowerMemmoveBase {
+public:
+  SPIRVLowerMemmoveBase() : Context(nullptr) {}
+
+  void LowerMemMoveInst(MemMoveInst &I);
+  bool expandMemMoveIntrinsicUses(Function &F);
+  bool runLowerMemmove(Module &M);
+
+private:
+  LLVMContext *Context;
+};
+
+class SPIRVLowerMemmovePass : public llvm::PassInfoMixin<SPIRVLowerMemmovePass>,
+                              public SPIRVLowerMemmoveBase {
+public:
+  llvm::PreservedAnalyses run(llvm::Module &M,
+                              llvm::ModuleAnalysisManager &MAM);
+};
+
+class SPIRVLowerMemmoveLegacy : public ModulePass,
+                                public SPIRVLowerMemmoveBase {
+public:
+  SPIRVLowerMemmoveLegacy();
+  bool runOnModule(Module &M) override;
+
+  static char ID;
+};
+
+} // namespace SPIRV
+
+#endif // SPIRV_SPIRVLOWERMEMMOVE_H
diff --git a/lib/SPIRV/SPIRVLowerOCLBlocks.cpp b/lib/SPIRV/SPIRVLowerOCLBlocks.cpp
index 59db06b1..61aad1f2 100644
--- a/lib/SPIRV/SPIRVLowerOCLBlocks.cpp
+++ b/lib/SPIRV/SPIRVLowerOCLBlocks.cpp
@@ -55,6 +55,7 @@
 //===----------------------------------------------------------------------===//
 #define DEBUG_TYPE "spv-lower-ocl-blocks"
 
+#include "SPIRVLowerOCLBlocks.h"
 #include "SPIRVInternal.h"
 
 #include "llvm/IR/Module.h"
@@ -71,58 +72,38 @@ static bool isBlockInvoke(Function &F) {
   return BlockInvokeRegex.match(F.getName());
 }
 
-class SPIRVLowerOCLBlocksBase {
+} // namespace
 
-public:
-  SPIRVLowerOCLBlocksBase() {}
+namespace SPIRV {
 
-  bool runLowerOCLBlocks(Module &M) {
-    bool Changed = false;
-    for (Function &F : M) {
-      if (!isBlockInvoke(F))
+bool SPIRVLowerOCLBlocksBase::runLowerOCLBlocks(Module &M) {
+  bool Changed = false;
+  for (Function &F : M) {
+    if (!isBlockInvoke(F))
+      continue;
+    for (User *U : F.users()) {
+      if (!isa<Constant>(U))
         continue;
-      for (User *U : F.users()) {
-        if (!isa<Constant>(U))
-          continue;
-        Constant *Null = Constant::getNullValue(U->getType());
-        if (U != Null) {
-          U->replaceAllUsesWith(Null);
-          Changed = true;
-        }
+      Constant *Null = Constant::getNullValue(U->getType());
+      if (U != Null) {
+        U->replaceAllUsesWith(Null);
+        Changed = true;
       }
     }
-    return Changed;
-  }
-};
-
-class SPIRVLowerOCLBlocksPass
-    : public llvm::PassInfoMixin<SPIRVLowerOCLBlocksPass>,
-      public SPIRVLowerOCLBlocksBase {
-public:
-  llvm::PreservedAnalyses run(llvm::Module &M,
-                              llvm::ModuleAnalysisManager &MAM) {
-    return runLowerOCLBlocks(M) ? llvm::PreservedAnalyses::none()
-                                : llvm::PreservedAnalyses::all();
-  }
-};
-
-class SPIRVLowerOCLBlocksLegacy : public ModulePass,
-                                  public SPIRVLowerOCLBlocksBase {
-public:
-  SPIRVLowerOCLBlocksLegacy() : ModulePass(ID) {}
-
-  bool runOnModule(Module &M) override { return runLowerOCLBlocks(M); }
-
-  StringRef getPassName() const override {
-    return "Lower OpenCL Blocks For SPIR-V";
   }
+  return Changed;
+}
 
-  static char ID;
-};
+llvm::PreservedAnalyses
+SPIRVLowerOCLBlocksPass::run(llvm::Module &M,
+                             llvm::ModuleAnalysisManager &MAM) {
+  return runLowerOCLBlocks(M) ? llvm::PreservedAnalyses::none()
+                              : llvm::PreservedAnalyses::all();
+}
 
 char SPIRVLowerOCLBlocksLegacy::ID = 0;
 
-} // namespace
+} // namespace SPIRV
 
 INITIALIZE_PASS(SPIRVLowerOCLBlocksLegacy, "spv-lower-ocl-blocks",
                 "Remove function pointers originating from OpenCL blocks",
diff --git a/lib/SPIRV/SPIRVLowerOCLBlocks.h b/lib/SPIRV/SPIRVLowerOCLBlocks.h
new file mode 100644
index 00000000..768da04b
--- /dev/null
+++ b/lib/SPIRV/SPIRVLowerOCLBlocks.h
@@ -0,0 +1,74 @@
+//===- SPIRVLowerOCLBlocks.h - OpenCL block lowering  --------*- C++ -*-===//
+//
+//                     The LLVM/SPIR-V Translator
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright (c) 2022 The Khronos Group Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of The Khronos Group, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SPIRV_SPIRVLOWEROCLBLOCKS_H
+#define SPIRV_SPIRVLOWEROCLBLOCKS_H
+
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+namespace SPIRV {
+
+class SPIRVLowerOCLBlocksBase {
+public:
+  SPIRVLowerOCLBlocksBase() {}
+
+  bool runLowerOCLBlocks(llvm::Module &M);
+};
+
+class SPIRVLowerOCLBlocksPass
+    : public llvm::PassInfoMixin<SPIRVLowerOCLBlocksPass>,
+      public SPIRVLowerOCLBlocksBase {
+public:
+  llvm::PreservedAnalyses run(llvm::Module &M,
+                              llvm::ModuleAnalysisManager &MAM);
+};
+
+class SPIRVLowerOCLBlocksLegacy : public llvm::ModulePass,
+                                  public SPIRVLowerOCLBlocksBase {
+public:
+  SPIRVLowerOCLBlocksLegacy() : ModulePass(ID) {}
+
+  bool runOnModule(llvm::Module &M) override { return runLowerOCLBlocks(M); }
+
+  llvm::StringRef getPassName() const override {
+    return "Lower OpenCL Blocks For SPIR-V";
+  }
+
+  static char ID;
+};
+
+} // namespace SPIRV
+
+#endif // SPIRV_SPIRVLOWEROCLBLOCKS_H
diff --git a/lib/SPIRV/SPIRVLowerSaddWithOverflow.cpp b/lib/SPIRV/SPIRVLowerSaddWithOverflow.cpp
index 19d60e1b..3695b024 100644
--- a/lib/SPIRV/SPIRVLowerSaddWithOverflow.cpp
+++ b/lib/SPIRV/SPIRVLowerSaddWithOverflow.cpp
@@ -39,6 +39,7 @@
 //===----------------------------------------------------------------------===//
 #define DEBUG_TYPE "spv-lower-llvm_sadd_with_overflow"
 
+#include "SPIRVLowerSaddWithOverflow.h"
 #include "LLVMSaddWithOverflow.h"
 
 #include "LLVMSPIRVLib.h"
@@ -48,10 +49,8 @@
 #include "llvm/IR/InstVisitor.h"
 #include "llvm/IR/IntrinsicInst.h"
 #include "llvm/IR/Module.h"
-#include "llvm/IR/PassManager.h"
 #include "llvm/IRReader/IRReader.h"
 #include "llvm/Linker/Linker.h"
-#include "llvm/Pass.h"
 #include "llvm/Support/SourceMgr.h"
 
 using namespace llvm;
@@ -59,103 +58,86 @@ using namespace SPIRV;
 
 namespace SPIRV {
 
-class SPIRVLowerSaddWithOverflowBase
-    : public InstVisitor<SPIRVLowerSaddWithOverflowBase> {
-public:
-  SPIRVLowerSaddWithOverflowBase() : Context(nullptr) {}
-  virtual ~SPIRVLowerSaddWithOverflowBase() {}
-  virtual void visitIntrinsicInst(CallInst &I) {
-    IntrinsicInst *II = dyn_cast<IntrinsicInst>(&I);
-    if (!II || II->getIntrinsicID() != Intrinsic::sadd_with_overflow)
-      return;
-
-    Function *IntrinsicFunc = I.getCalledFunction();
-    assert(IntrinsicFunc && "Missing function");
-    StringRef IntrinsicName = IntrinsicFunc->getName();
-    std::string FuncName = "llvm_sadd_with_overflow_i";
-    if (IntrinsicName.endswith(".i16"))
-      FuncName += "16";
-    else if (IntrinsicName.endswith(".i32"))
-      FuncName += "32";
-    else if (IntrinsicName.endswith(".i64"))
-      FuncName += "64";
-    else {
-      assert(false &&
-             "Unsupported overloading of llvm.sadd.with.overflow intrinsic");
-      return;
-    }
-
-    // Redirect @llvm.sadd.with.overflow.* call to the function we have in
-    // the loaded module @llvm_sadd_with_overflow_*
-    Function *F = Mod->getFunction(FuncName);
-    if (F) { // This function is already linked in.
-      I.setCalledFunction(F);
-      return;
-    }
-    FunctionCallee FC = Mod->getOrInsertFunction(FuncName, I.getFunctionType());
-    I.setCalledFunction(FC);
-
-    // Read LLVM IR with the intrinsic's implementation
-    SMDiagnostic Err;
-    auto MB = MemoryBuffer::getMemBuffer(LLVMSaddWithOverflow);
-    auto SaddWithOverflowModule =
-        parseIR(MB->getMemBufferRef(), Err, *Context,
-                [&](StringRef) { return Mod->getDataLayoutStr(); });
-    if (!SaddWithOverflowModule) {
-      std::string ErrMsg;
-      raw_string_ostream ErrStream(ErrMsg);
-      Err.print("", ErrStream);
-      SPIRVErrorLog EL;
-      EL.checkError(false, SPIRVEC_InvalidLlvmModule, ErrMsg);
-      return;
-    }
-
-    // Link in the intrinsic's implementation.
-    if (!Linker::linkModules(*Mod, std::move(SaddWithOverflowModule),
-                             Linker::LinkOnlyNeeded))
-      TheModuleIsModified = true;
+void SPIRVLowerSaddWithOverflowBase::visitIntrinsicInst(CallInst &I) {
+  IntrinsicInst *II = dyn_cast<IntrinsicInst>(&I);
+  if (!II || II->getIntrinsicID() != Intrinsic::sadd_with_overflow)
+    return;
+
+  Function *IntrinsicFunc = I.getCalledFunction();
+  assert(IntrinsicFunc && "Missing function");
+  StringRef IntrinsicName = IntrinsicFunc->getName();
+  std::string FuncName = "llvm_sadd_with_overflow_i";
+  if (IntrinsicName.endswith(".i16"))
+    FuncName += "16";
+  else if (IntrinsicName.endswith(".i32"))
+    FuncName += "32";
+  else if (IntrinsicName.endswith(".i64"))
+    FuncName += "64";
+  else {
+    assert(false &&
+           "Unsupported overloading of llvm.sadd.with.overflow intrinsic");
+    return;
   }
 
-  bool runLowerSaddWithOverflow(Module &M) {
-    Context = &M.getContext();
-    Mod = &M;
-    visit(M);
-
-    verifyRegularizationPass(M, "SPIRVLowerSaddWithOverflow");
-    return TheModuleIsModified;
-  }
-
-private:
-  LLVMContext *Context;
-  Module *Mod;
-  bool TheModuleIsModified = false;
-};
-
-class SPIRVLowerSaddWithOverflowPass
-    : public llvm::PassInfoMixin<SPIRVLowerSaddWithOverflowPass>,
-      public SPIRVLowerSaddWithOverflowBase {
-public:
-  llvm::PreservedAnalyses run(llvm::Module &M,
-                              llvm::ModuleAnalysisManager &MAM) {
-    return runLowerSaddWithOverflow(M) ? llvm::PreservedAnalyses::none()
-                                       : llvm::PreservedAnalyses::all();
+  // Redirect @llvm.sadd.with.overflow.* call to the function we have in
+  // the loaded module @llvm_sadd_with_overflow_*
+  Function *F = Mod->getFunction(FuncName);
+  if (F) { // This function is already linked in.
+    I.setCalledFunction(F);
+    return;
   }
-};
-
-class SPIRVLowerSaddWithOverflowLegacy : public ModulePass,
-                                         public SPIRVLowerSaddWithOverflowBase {
-public:
-  SPIRVLowerSaddWithOverflowLegacy() : ModulePass(ID) {
-    initializeSPIRVLowerSaddWithOverflowLegacyPass(
-        *PassRegistry::getPassRegistry());
+  FunctionCallee FC = Mod->getOrInsertFunction(FuncName, I.getFunctionType());
+  I.setCalledFunction(FC);
+
+  // Read LLVM IR with the intrinsic's implementation
+  SMDiagnostic Err;
+  auto MB = MemoryBuffer::getMemBuffer(LLVMSaddWithOverflow);
+  auto SaddWithOverflowModule =
+      parseIR(MB->getMemBufferRef(), Err, *Context,
+              [&](StringRef) { return Mod->getDataLayoutStr(); });
+  if (!SaddWithOverflowModule) {
+    std::string ErrMsg;
+    raw_string_ostream ErrStream(ErrMsg);
+    Err.print("", ErrStream);
+    SPIRVErrorLog EL;
+    EL.checkError(false, SPIRVEC_InvalidLlvmModule, ErrMsg);
+    return;
   }
 
-  bool runOnModule(Module &M) override { return runLowerSaddWithOverflow(M); }
+  // Link in the intrinsic's implementation.
+  if (!Linker::linkModules(*Mod, std::move(SaddWithOverflowModule),
+                           Linker::LinkOnlyNeeded))
+    TheModuleIsModified = true;
+}
 
-  static char ID;
-};
+bool SPIRVLowerSaddWithOverflowBase::runLowerSaddWithOverflow(Module &M) {
+  Context = &M.getContext();
+  Mod = &M;
+  visit(M);
+
+  verifyRegularizationPass(M, "SPIRVLowerSaddWithOverflow");
+  return TheModuleIsModified;
+}
+
+llvm::PreservedAnalyses
+SPIRVLowerSaddWithOverflowPass::run(llvm::Module &M,
+                                    llvm::ModuleAnalysisManager &MAM) {
+  return runLowerSaddWithOverflow(M) ? llvm::PreservedAnalyses::none()
+                                     : llvm::PreservedAnalyses::all();
+}
+
+SPIRVLowerSaddWithOverflowLegacy::SPIRVLowerSaddWithOverflowLegacy()
+    : ModulePass(ID) {
+  initializeSPIRVLowerSaddWithOverflowLegacyPass(
+      *PassRegistry::getPassRegistry());
+}
+
+bool SPIRVLowerSaddWithOverflowLegacy::runOnModule(Module &M) {
+  return runLowerSaddWithOverflow(M);
+}
 
 char SPIRVLowerSaddWithOverflowLegacy::ID = 0;
+
 } // namespace SPIRV
 
 INITIALIZE_PASS(SPIRVLowerSaddWithOverflowLegacy,
diff --git a/lib/SPIRV/SPIRVLowerSaddWithOverflow.h b/lib/SPIRV/SPIRVLowerSaddWithOverflow.h
new file mode 100644
index 00000000..9e8759a1
--- /dev/null
+++ b/lib/SPIRV/SPIRVLowerSaddWithOverflow.h
@@ -0,0 +1,81 @@
+//===- SPIRVLowerSaddWithOverflowBase.h - sadd lowering  --------*- C++ -*-===//
+//
+//                     The LLVM/SPIR-V Translator
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright (c) 2022 The Khronos Group Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of The Khronos Group, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SPIRV_SPIRVLOWERSADDWITHOVERFLOW_H
+#define SPIRV_SPIRVLOWERSADDWITHOVERFLOW_H
+
+#include "llvm/IR/InstVisitor.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+namespace SPIRV {
+
+class SPIRVLowerSaddWithOverflowBase
+    : public InstVisitor<SPIRVLowerSaddWithOverflowBase> {
+public:
+  SPIRVLowerSaddWithOverflowBase() : Context(nullptr) {}
+  virtual ~SPIRVLowerSaddWithOverflowBase() {}
+  virtual void visitIntrinsicInst(CallInst &I);
+
+  bool runLowerSaddWithOverflow(Module &M);
+
+private:
+  LLVMContext *Context;
+  Module *Mod;
+  bool TheModuleIsModified = false;
+};
+
+class SPIRVLowerSaddWithOverflowPass
+    : public llvm::PassInfoMixin<SPIRVLowerSaddWithOverflowPass>,
+      public SPIRVLowerSaddWithOverflowBase {
+public:
+  llvm::PreservedAnalyses run(llvm::Module &M,
+                              llvm::ModuleAnalysisManager &MAM);
+};
+
+class SPIRVLowerSaddWithOverflowLegacy : public ModulePass,
+                                         public SPIRVLowerSaddWithOverflowBase {
+public:
+  SPIRVLowerSaddWithOverflowLegacy();
+
+  bool runOnModule(Module &M) override;
+
+  static char ID;
+};
+
+} // namespace SPIRV
+
+#endif // SPIRV_SPIRVLOWERSADDWITHOVERFLOW_H
diff --git a/lib/SPIRV/SPIRVRegularizeLLVM.cpp b/lib/SPIRV/SPIRVRegularizeLLVM.cpp
index 8e86b9aa..2166c127 100644
--- a/lib/SPIRV/SPIRVRegularizeLLVM.cpp
+++ b/lib/SPIRV/SPIRVRegularizeLLVM.cpp
@@ -37,6 +37,7 @@
 //===----------------------------------------------------------------------===//
 #define DEBUG_TYPE "spvregular"
 
+#include "SPIRVRegularizeLLVM.h"
 #include "OCLUtil.h"
 #include "SPIRVInternal.h"
 #include "SPIRVMDWalker.h"
@@ -45,11 +46,9 @@
 #include "llvm/ADT/StringExtras.h" // llvm::isDigit
 #include "llvm/CodeGen/IntrinsicLowering.h"
 #include "llvm/Demangle/Demangle.h"
-#include "llvm/IR/InstVisitor.h"
-#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/Module.h"
 #include "llvm/IR/Operator.h"
-#include "llvm/IR/PassManager.h"
-#include "llvm/Pass.h"
 #include "llvm/Support/Debug.h"
 #include "llvm/Transforms/Utils/LowerMemIntrinsics.h" // expandMemSetAsLoop()
 
@@ -65,99 +64,6 @@ namespace SPIRV {
 static bool SPIRVDbgSaveRegularizedModule = false;
 static std::string RegularizedModuleTmpFile = "regularized.bc";
 
-class SPIRVRegularizeLLVMBase {
-public:
-  SPIRVRegularizeLLVMBase() : M(nullptr), Ctx(nullptr) {}
-
-  bool runRegularizeLLVM(Module &M);
-  // Lower functions
-  bool regularize();
-
-  // SPIR-V disallows functions being entrypoints and called
-  // LLVM doesn't. This adds a wrapper around the entry point
-  // that later SPIR-V writer renames.
-  void addKernelEntryPoint(Module *M);
-
-  /// Some LLVM intrinsics that have no SPIR-V counterpart may be wrapped in
-  /// @spirv.llvm_intrinsic_* function. During reverse translation from SPIR-V
-  /// to LLVM IR we can detect this @spirv.llvm_intrinsic_* function and
-  /// replace it with @llvm.intrinsic.* back.
-  void lowerIntrinsicToFunction(IntrinsicInst *Intrinsic);
-
-  /// No SPIR-V counterpart for @llvm.fshl.*(@llvm.fshr.*) intrinsic. It will be
-  /// lowered to a newly generated @spirv.llvm_fshl_*(@spirv.llvm_fshr_*)
-  /// function.
-  ///
-  /// Conceptually, FSHL (FSHR):
-  /// 1. concatenates the ints, the first one being the more significant;
-  /// 2. performs a left (right) shift-rotate on the resulting doubled-sized
-  /// int;
-  /// 3. returns the most (least) significant bits of the shift-rotate result,
-  ///    the number of bits being equal to the size of the original integers.
-  /// If FSHL (FSHR) operates on a vector type instead, the same operations are
-  /// performed for each set of corresponding vector elements.
-  ///
-  /// The actual implementation algorithm will be slightly different for
-  /// simplification purposes.
-  void lowerFunnelShift(IntrinsicInst *FSHIntrinsic);
-
-  void lowerUMulWithOverflow(IntrinsicInst *UMulIntrinsic);
-  void buildUMulWithOverflowFunc(Function *UMulFunc);
-
-  // For some cases Clang emits VectorExtractDynamic as:
-  // void @_Z28__spirv_VectorExtractDynamic(<Ty>* sret(<Ty>), jointMatrix, idx);
-  // Instead of:
-  // <Ty> @_Z28__spirv_VectorExtractDynamic(JointMatrix, Idx);
-  // And VectorInsertDynamic as:
-  // @_Z27__spirv_VectorInsertDynamic(jointMatrix, <Ty>* byval(<Ty>), idx);
-  // Instead of:
-  // @_Z27__spirv_VectorInsertDynamic(jointMatrix, <Ty>, idx)
-  // Need to add additional GEP, store and load instructions and mutate called
-  // function to avoid translation failures
-  void expandSYCLTypeUsing(Module *M);
-  void expandVEDWithSYCLTypeSRetArg(Function *F);
-  void expandVIDWithSYCLTypeByValComp(Function *F);
-
-  // According to the specification, the operands of a shift instruction must be
-  // a scalar/vector of integer. When LLVM-IR contains a shift instruction with
-  // i1 operands, they are treated as a bool. We need to extend them to i32 to
-  // comply with the specification. For example: "%shift = lshr i1 0, 1";
-  // The bit instruction should be changed to the extended version
-  // "%shift = lshr i32 0, 1" so the args are treated as int operands.
-  Value *extendBitInstBoolArg(Instruction *OldInst);
-
-  static std::string lowerLLVMIntrinsicName(IntrinsicInst *II);
-  void adaptStructTypes(StructType *ST);
-  static char ID;
-
-private:
-  Module *M;
-  LLVMContext *Ctx;
-};
-
-class SPIRVRegularizeLLVMPass
-    : public llvm::PassInfoMixin<SPIRVRegularizeLLVMPass>,
-      public SPIRVRegularizeLLVMBase {
-public:
-  llvm::PreservedAnalyses run(llvm::Module &M,
-                              llvm::ModuleAnalysisManager &MAM) {
-    return runRegularizeLLVM(M) ? llvm::PreservedAnalyses::none()
-                                : llvm::PreservedAnalyses::all();
-  }
-};
-
-class SPIRVRegularizeLLVMLegacy : public ModulePass,
-                                  public SPIRVRegularizeLLVMBase {
-public:
-  SPIRVRegularizeLLVMLegacy() : ModulePass(ID) {
-    initializeSPIRVRegularizeLLVMLegacyPass(*PassRegistry::getPassRegistry());
-  }
-
-  bool runOnModule(Module &M) override;
-
-  static char ID;
-};
-
 char SPIRVRegularizeLLVMLegacy::ID = 0;
 
 bool SPIRVRegularizeLLVMLegacy::runOnModule(Module &Module) {
diff --git a/lib/SPIRV/SPIRVRegularizeLLVM.h b/lib/SPIRV/SPIRVRegularizeLLVM.h
new file mode 100644
index 00000000..7215b393
--- /dev/null
+++ b/lib/SPIRV/SPIRVRegularizeLLVM.h
@@ -0,0 +1,143 @@
+//=- SPIRVRegularizeLLVM.h - LLVM Module regularization pass -*- C++ -*-=//
+//
+//                     The LLVM/SPIR-V Translator
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright (c) 2022 The Khronos Group Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of The Khronos Group, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SPIRV_SPIRVREGULARIZELLVM_H
+#define SPIRV_SPIRVREGULARIZELLVM_H
+
+#include "SPIRVInternal.h"
+
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+namespace SPIRV {
+
+class SPIRVRegularizeLLVMBase {
+public:
+  SPIRVRegularizeLLVMBase() : M(nullptr), Ctx(nullptr) {}
+
+  bool runRegularizeLLVM(Module &M);
+  // Lower functions
+  bool regularize();
+
+  // SPIR-V disallows functions being entrypoints and called
+  // LLVM doesn't. This adds a wrapper around the entry point
+  // that later SPIR-V writer renames.
+  void addKernelEntryPoint(Module *M);
+
+  /// Some LLVM intrinsics that have no SPIR-V counterpart may be wrapped in
+  /// @spirv.llvm_intrinsic_* function. During reverse translation from SPIR-V
+  /// to LLVM IR we can detect this @spirv.llvm_intrinsic_* function and
+  /// replace it with @llvm.intrinsic.* back.
+  void lowerIntrinsicToFunction(IntrinsicInst *Intrinsic);
+
+  /// No SPIR-V counterpart for @llvm.fshl.*(@llvm.fshr.*) intrinsic. It will be
+  /// lowered to a newly generated @spirv.llvm_fshl_*(@spirv.llvm_fshr_*)
+  /// function.
+  ///
+  /// Conceptually, FSHL (FSHR):
+  /// 1. concatenates the ints, the first one being the more significant;
+  /// 2. performs a left (right) shift-rotate on the resulting doubled-sized
+  /// int;
+  /// 3. returns the most (least) significant bits of the shift-rotate result,
+  ///    the number of bits being equal to the size of the original integers.
+  /// If FSHL (FSHR) operates on a vector type instead, the same operations are
+  /// performed for each set of corresponding vector elements.
+  ///
+  /// The actual implementation algorithm will be slightly different for
+  /// simplification purposes.
+  void lowerFunnelShift(IntrinsicInst *FSHIntrinsic);
+
+  void lowerUMulWithOverflow(IntrinsicInst *UMulIntrinsic);
+  void buildUMulWithOverflowFunc(Function *UMulFunc);
+
+  // For some cases Clang emits VectorExtractDynamic as:
+  // void @_Z28__spirv_VectorExtractDynamic(<Ty>* sret(<Ty>), jointMatrix, idx);
+  // Instead of:
+  // <Ty> @_Z28__spirv_VectorExtractDynamic(JointMatrix, Idx);
+  // And VectorInsertDynamic as:
+  // @_Z27__spirv_VectorInsertDynamic(jointMatrix, <Ty>* byval(<Ty>), idx);
+  // Instead of:
+  // @_Z27__spirv_VectorInsertDynamic(jointMatrix, <Ty>, idx)
+  // Need to add additional GEP, store and load instructions and mutate called
+  // function to avoid translation failures
+  void expandSYCLTypeUsing(Module *M);
+  void expandVEDWithSYCLTypeSRetArg(Function *F);
+  void expandVIDWithSYCLTypeByValComp(Function *F);
+
+  // According to the specification, the operands of a shift instruction must be
+  // a scalar/vector of integer. When LLVM-IR contains a shift instruction with
+  // i1 operands, they are treated as a bool. We need to extend them to i32 to
+  // comply with the specification. For example: "%shift = lshr i1 0, 1";
+  // The bit instruction should be changed to the extended version
+  // "%shift = lshr i32 0, 1" so the args are treated as int operands.
+  Value *extendBitInstBoolArg(Instruction *OldInst);
+
+  static std::string lowerLLVMIntrinsicName(IntrinsicInst *II);
+  void adaptStructTypes(StructType *ST);
+  static char ID;
+
+private:
+  Module *M;
+  LLVMContext *Ctx;
+};
+
+class SPIRVRegularizeLLVMPass
+    : public llvm::PassInfoMixin<SPIRVRegularizeLLVMPass>,
+      public SPIRVRegularizeLLVMBase {
+public:
+  llvm::PreservedAnalyses run(llvm::Module &M,
+                              llvm::ModuleAnalysisManager &MAM) {
+    return runRegularizeLLVM(M) ? llvm::PreservedAnalyses::none()
+                                : llvm::PreservedAnalyses::all();
+  }
+};
+
+class SPIRVRegularizeLLVMLegacy : public ModulePass,
+                                  public SPIRVRegularizeLLVMBase {
+public:
+  SPIRVRegularizeLLVMLegacy() : ModulePass(ID) {
+    initializeSPIRVRegularizeLLVMLegacyPass(*PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override;
+
+  static char ID;
+};
+
+} // namespace SPIRV
+
+#endif // SPIRV_SPIRVREGULARIZELLVM_H
-- 
2.20.1

